

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>ChIP-seq Modules – HOMER &mdash; Omics Pipe v1.0.16 Documentation</title>
    
    <link rel="stylesheet" href="_static/cloud.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="//fonts.googleapis.com/css?family=Noticia+Text|Open+Sans|Droid+Sans+Mono" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.0.16',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="_static/jquery.cookie.js"></script>
    <script type="text/javascript" src="_static/cloud.js"></script>
    <link rel="author" title="About these documents" href="about.html" />
    <link rel="copyright" title="Copyright" href="copyright.html" />
    <link rel="top" title="Omics Pipe v1.0.16 Documentation" href="index.html" />
    <link rel="next" title="ChIP-seq Modules – MACS" href="ChIPseq_modules_MACS.html" />
    <link rel="prev" title="miRNA-seq Count Based Modules" href="miRNAseq_counts_modules.html" /> 
        <meta name="viewport" content="width=device-width, initial-scale=1">
  </head>
  <body>
    <div class="relbar-top">
        
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> &nbsp; &nbsp;</li>
        <li class="right" >
          <a href="ChIPseq_modules_MACS.html" title="ChIP-seq Modules – MACS"
             accesskey="N">next</a> &nbsp; &nbsp;</li>
        <li class="right" >
          <a href="miRNAseq_counts_modules.html" title="miRNA-seq Count Based Modules"
             accesskey="P">previous</a> &nbsp; &nbsp;</li>
    <li><a href="contents.html">Omics Pipe 1.0.16 Documentation</a> &raquo;</li>
 
      </ul>
    </div>
    </div>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="chip-seq-modules-homer">
<h1>ChIP-seq Modules &#8211; HOMER<a class="headerlink" href="#chip-seq-modules-homer" title="Permalink to this headline">¶</a></h1>
<div class="html-toggle section" id="module-omics_pipe.modules.fastqc">
<span id="fastqc"></span><h2>FASTQC<a class="headerlink" href="#module-omics_pipe.modules.fastqc" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="omics_pipe.modules.fastqc.fastqc">
<tt class="descclassname">omics_pipe.modules.fastqc.</tt><tt class="descname">fastqc</tt><big>(</big><em>sample</em>, <em>fastqc_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/fastqc.html#fastqc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.fastqc.fastqc" title="Permalink to this definition">¶</a></dt>
<dd><p>QC check of raw .fastq files using FASTQC.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.fastq file</dd>
<dt>output: </dt>
<dd>folder and zipped folder containing html, txt and image files</dd>
<dt>citation: </dt>
<dd>Babraham Bioinformatics</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">http://www.bioinformatics.babraham.ac.uk/projects/fastqc/</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">RAW_DATA_DIR:</p>
<p>QC_PATH:</p>
<p class="last">FASTQC_VERSION:</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="html-toggle section" id="module-omics_pipe.modules.ChIP_trim">
<span id="chip-trim"></span><h2>ChIP trim<a class="headerlink" href="#module-omics_pipe.modules.ChIP_trim" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="omics_pipe.modules.ChIP_trim.ChIP_trim">
<tt class="descclassname">omics_pipe.modules.ChIP_trim.</tt><tt class="descname">ChIP_trim</tt><big>(</big><em>sample</em>, <em>ChIP_trim_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/ChIP_trim.html#ChIP_trim"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.ChIP_trim.ChIP_trim" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs Homer Tools to trim adapters from .fastq files.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.fastq file</dd>
<dt>output: </dt>
<dd>.fastq file</dd>
<dt>citation: </dt>
<dd>Heinz S, Benner C, Spann N, Bertolino E et al. Simple Combinations of Lineage-Determining Transcription Factors Prime cis-Regulatory Elements Required for Macrophage and B Cell Identities. Mol Cell 2010 May 28;38(4):576-589. PMID: 20513432</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://homer.salk.edu/homer/">http://homer.salk.edu/homer/</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">ENDS:</p>
<p>RAW_DATA_DIR:</p>
<p>HOMER_TRIM_OPTIONS:</p>
<p>TRIMMED_DATA_PATH:</p>
<p class="last">HOMER_VERSION:</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="html-toggle section" id="module-omics_pipe.modules.bowtie">
<span id="bowtie"></span><h2>Bowtie<a class="headerlink" href="#module-omics_pipe.modules.bowtie" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="omics_pipe.modules.bowtie.bowtie">
<tt class="descclassname">omics_pipe.modules.bowtie.</tt><tt class="descname">bowtie</tt><big>(</big><em>sample</em>, <em>bowtie_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/bowtie.html#bowtie"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.bowtie.bowtie" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs Bowtie to align .fastq files.</p>
<dl class="docutils">
<dt>input:</dt>
<dd>.fastq file</dd>
<dt>output:</dt>
<dd>sample.bam</dd>
<dt>citation: </dt>
<dd>Langmead B, Trapnell C, Pop M, Salzberg SL. Ultrafast and memory-efficient alignment of short DNA sequences to the human genome. Genome Biology 10:R25</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://bowtie-bio.sourceforge.net/index.shtml">http://bowtie-bio.sourceforge.net/index.shtml</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">ENDS:</p>
<p>TRIMMED_DATA_PATH:</p>
<p>BOWTIE_OPTIONS:</p>
<p>BOWTIE_INDEX:</p>
<p>BOWTIE_RESULTS:</p>
<p>BOWTIE_VERSION:</p>
<p>SAMTOOLS_VERSION:</p>
<p class="last">BEDTOOLS_VERSION:</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="html-toggle section" id="module-omics_pipe.modules.read_density">
<span id="read-density-homer"></span><h2>Read Density -HOMER<a class="headerlink" href="#module-omics_pipe.modules.read_density" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="omics_pipe.modules.read_density.read_density">
<tt class="descclassname">omics_pipe.modules.read_density.</tt><tt class="descname">read_density</tt><big>(</big><em>sample</em>, <em>read_density_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/read_density.html#read_density"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.read_density.read_density" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs HOMER to visualize read density from ChIPseq data.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.bam file</dd>
<dt>output: </dt>
<dd>.txt file</dd>
<dt>citation: </dt>
<dd>Heinz S, Benner C, Spann N, Bertolino E et al. Simple Combinations of Lineage-Determining Transcription Factors Prime cis-Regulatory Elements Required for Macrophage and B Cell Identities. Mol Cell 2010 May 28;38(4):576-589. PMID: 20513432</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://homer.salk.edu/homer/">http://homer.salk.edu/homer/</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">BOWTIE_RESULTS:</p>
<p>CHROM_SIZES:</p>
<p>HOMER_RESULTS:</p>
<p>HOMER_VERSION:</p>
<p class="last">TEMP_DIR:</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="html-toggle section" id="peak-detection-homer">
<h2>Peak Detection - HOMER<a class="headerlink" href="#peak-detection-homer" title="Permalink to this headline">¶</a></h2>
<span class="target" id="module-omics_pipe.modules.homer_peaks"></span><dl class="function">
<dt id="omics_pipe.modules.homer_peaks.homer_peaks">
<tt class="descclassname">omics_pipe.modules.homer_peaks.</tt><tt class="descname">homer_peaks</tt><big>(</big><em>step</em>, <em>homer_peaks_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/homer_peaks.html#homer_peaks"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.homer_peaks.homer_peaks" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs HOMER to call peaks from ChIPseq data.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.tag input file</dd>
<dt>output: </dt>
<dd>.txt file</dd>
<dt>citation: </dt>
<dd>Heinz S, Benner C, Spann N, Bertolino E et al. Simple Combinations of Lineage-Determining Transcription Factors Prime cis-Regulatory Elements Required for Macrophage and B Cell Identities. Mol Cell 2010 May 28;38(4):576-589. PMID: 20513432</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://homer.salk.edu/homer/">http://homer.salk.edu/homer/</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">PAIR_LIST:</p>
<p>HOMER_RESULTS:</p>
<p>HOMER_PEAKS_OPTIONS:</p>
<p>HOMER_VERSION:</p>
<p class="last">TEMP_DIR:</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="html-toggle section" id="peak-annotation-visualization-homer">
<h2>Peak Annotation &amp; Visualization - HOMER<a class="headerlink" href="#peak-annotation-visualization-homer" title="Permalink to this headline">¶</a></h2>
<span class="target" id="module-omics_pipe.modules.peak_track"></span><dl class="function">
<dt id="omics_pipe.modules.peak_track.peak_track">
<tt class="descclassname">omics_pipe.modules.peak_track.</tt><tt class="descname">peak_track</tt><big>(</big><em>step</em>, <em>peak_track_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/peak_track.html#peak_track"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.peak_track.peak_track" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs HOMER to create peak track from ChIPseq data.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.tag input file</dd>
<dt>output: </dt>
<dd>.txt file</dd>
<dt>citation: </dt>
<dd>Heinz S, Benner C, Spann N, Bertolino E et al. Simple Combinations of Lineage-Determining Transcription Factors Prime cis-Regulatory Elements Required for Macrophage and B Cell Identities. Mol Cell 2010 May 28;38(4):576-589. PMID: 20513432</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://homer.salk.edu/homer/">http://homer.salk.edu/homer/</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">PAIR_LIST:</p>
<p>HOMER_RESULTS:</p>
<p>HOMER_VERSION:</p>
<p class="last">TEMP_DIR:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.annotate_peaks"></span><dl class="function">
<dt id="omics_pipe.modules.annotate_peaks.annotate_peaks">
<tt class="descclassname">omics_pipe.modules.annotate_peaks.</tt><tt class="descname">annotate_peaks</tt><big>(</big><em>step</em>, <em>annotate_peaks_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/annotate_peaks.html#annotate_peaks"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.annotate_peaks.annotate_peaks" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs HOMER to annotate peak track from ChIPseq data.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.tag input file</dd>
<dt>output: </dt>
<dd>.txt file</dd>
<dt>citation:</dt>
<dd>Heinz S, Benner C, Spann N, Bertolino E et al. Simple Combinations of Lineage-Determining Transcription Factors Prime cis-Regulatory Elements Required for Macrophage and B Cell Identities. Mol Cell 2010 May 28;38(4):576-589. PMID: 20513432</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://homer.salk.edu/homer/">http://homer.salk.edu/homer/</a></dd>
<dt>parameters from parameters file:</dt>
<dd><p class="first">PAIR_LIST:</p>
<p>HOMER_RESULTS:</p>
<p>HOMER_VERSION:</p>
<p>TEMP_DIR:</p>
<p>HOMER_GENOME:</p>
<p class="last">HOMER_ANNOTATE_OPTIONS:</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="html-toggle section" id="module-omics_pipe.modules.find_motifs">
<span id="find-motifs-homer"></span><h2>Find Motifs - HOMER<a class="headerlink" href="#module-omics_pipe.modules.find_motifs" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="omics_pipe.modules.find_motifs.find_motifs">
<tt class="descclassname">omics_pipe.modules.find_motifs.</tt><tt class="descname">find_motifs</tt><big>(</big><em>step</em>, <em>find_motifs_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/find_motifs.html#find_motifs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.find_motifs.find_motifs" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs HOMER to find motifs from ChIPseq data.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.txt peak file from Homer</dd>
<dt>output: </dt>
<dd>.txt file</dd>
<dt>citation: </dt>
<dd>Heinz S, Benner C, Spann N, Bertolino E et al. Simple Combinations of Lineage-Determining Transcription Factors Prime cis-Regulatory Elements Required for Macrophage and B Cell Identities. Mol Cell 2010 May 28;38(4):576-589. PMID: 20513432</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://homer.salk.edu/homer/">http://homer.salk.edu/homer/</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">PAIR_LIST:</p>
<p>HOMER_RESULTS:</p>
<p>HOMER_VERSION:</p>
<p>TEMP_DIR:</p>
<p>HOMER_GENOME:</p>
<p class="last">HOMER_MOTIFS_OPTIONS:</p>
</dd>
</dl>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
        <p class="logo"><a href="contents.html" title="contents">
          <img class="logo" src="_static/logo.jpg" alt="Logo"/>
        </a></p><div class="sphinxlocaltoc">
    <h3><a href="contents.html">Page contents</a></h3>
    <ul>
<li><a class="reference internal" href="#">ChIP-seq Modules &#8211; HOMER</a><ul>
<li><a class="reference internal" href="#module-omics_pipe.modules.fastqc">FASTQC</a></li>
<li><a class="reference internal" href="#module-omics_pipe.modules.ChIP_trim">ChIP trim</a></li>
<li><a class="reference internal" href="#module-omics_pipe.modules.bowtie">Bowtie</a></li>
<li><a class="reference internal" href="#module-omics_pipe.modules.read_density">Read Density -HOMER</a></li>
<li><a class="reference internal" href="#peak-detection-homer">Peak Detection - HOMER</a></li>
<li><a class="reference internal" href="#peak-annotation-visualization-homer">Peak Annotation &amp; Visualization - HOMER</a></li>
<li><a class="reference internal" href="#module-omics_pipe.modules.find_motifs">Find Motifs - HOMER</a></li>
</ul>
</li>
</ul>

  </div>
  <div class="sphinxprev">
    <h4>Previous page</h4>
    <p class="topless"><a href="miRNAseq_counts_modules.html"
                          title="Previous page">&larr; miRNA-seq Count Based Modules</a></p>
  </div>
  <div class="sphinxnext">
    <h4>Next page</h4>
    <p class="topless"><a href="ChIPseq_modules_MACS.html"
                          title="Next page">&rarr; ChIP-seq Modules &#8211; MACS</a></p>
  </div>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/ChIPseq_modules_HOMER.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="relbar-bottom">
        
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> &nbsp; &nbsp;</li>
        <li class="right" >
          <a href="ChIPseq_modules_MACS.html" title="ChIP-seq Modules – MACS"
             >next</a> &nbsp; &nbsp;</li>
        <li class="right" >
          <a href="miRNAseq_counts_modules.html" title="miRNA-seq Count Based Modules"
             >previous</a> &nbsp; &nbsp;</li>
    <li><a href="contents.html">Omics Pipe 1.0.16 Documentation</a> &raquo;</li>
 
      </ul>
    </div>
    </div>

    <div class="footer">
        &copy; <a href="copyright.html">Copyright</a> 2014, Kathleen Fisch, Ph.D..
      Last updated on Jul 16, 2014.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.2.
    </div>
    <!-- cloud_sptheme 1.4 -->
  </body>
</html>