

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Breast Cancer Personalized Genomics Report- RNAseq &mdash; Omics Pipe v1.0.16 Documentation</title>
    
    <link rel="stylesheet" href="_static/cloud.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="//fonts.googleapis.com/css?family=Noticia+Text|Open+Sans|Droid+Sans+Mono" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '1.0.16',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="_static/jquery.cookie.js"></script>
    <script type="text/javascript" src="_static/cloud.js"></script>
    <link rel="author" title="About these documents" href="about.html" />
    <link rel="copyright" title="Copyright" href="copyright.html" />
    <link rel="top" title="Omics Pipe v1.0.16 Documentation" href="index.html" />
    <link rel="next" title="TCGA Reanalysis Pipeline - RNAseq" href="RNAseq_cancer_modules_TCGA.html" />
    <link rel="prev" title="RNA-seq Count Based Modules" href="RNAseq_counts_modules.html" /> 
        <meta name="viewport" content="width=device-width, initial-scale=1">
  </head>
  <body>
    <div class="relbar-top">
        
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> &nbsp; &nbsp;</li>
        <li class="right" >
          <a href="RNAseq_cancer_modules_TCGA.html" title="TCGA Reanalysis Pipeline - RNAseq"
             accesskey="N">next</a> &nbsp; &nbsp;</li>
        <li class="right" >
          <a href="RNAseq_counts_modules.html" title="RNA-seq Count Based Modules"
             accesskey="P">previous</a> &nbsp; &nbsp;</li>
    <li><a href="contents.html">Omics Pipe 1.0.16 Documentation</a> &raquo;</li>
 
      </ul>
    </div>
    </div>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="breast-cancer-personalized-genomics-report-rnaseq">
<h1>Breast Cancer Personalized Genomics Report- RNAseq<a class="headerlink" href="#breast-cancer-personalized-genomics-report-rnaseq" title="Permalink to this headline">¶</a></h1>
<p>Modules included in the RNAseq Cancer pipeline.</p>
<div class="html-toggle section" id="module-omics_pipe.modules.fastqc">
<span id="fastqc"></span><h2>FASTQC<a class="headerlink" href="#module-omics_pipe.modules.fastqc" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="omics_pipe.modules.fastqc.fastqc">
<tt class="descclassname">omics_pipe.modules.fastqc.</tt><tt class="descname">fastqc</tt><big>(</big><em>sample</em>, <em>fastqc_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/fastqc.html#fastqc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.fastqc.fastqc" title="Permalink to this definition">¶</a></dt>
<dd><p>QC check of raw .fastq files using FASTQC.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.fastq file</dd>
<dt>output: </dt>
<dd>folder and zipped folder containing html, txt and image files</dd>
<dt>citation: </dt>
<dd>Babraham Bioinformatics</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">http://www.bioinformatics.babraham.ac.uk/projects/fastqc/</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">RAW_DATA_DIR:</p>
<p>QC_PATH:</p>
<p class="last">FASTQC_VERSION:</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="html-toggle section" id="star-aligner">
<h2>STAR Aligner<a class="headerlink" href="#star-aligner" title="Permalink to this headline">¶</a></h2>
<span class="target" id="module-omics_pipe.modules.star"></span><dl class="function">
<dt id="omics_pipe.modules.star.star">
<tt class="descclassname">omics_pipe.modules.star.</tt><tt class="descname">star</tt><big>(</big><em>sample</em>, <em>star_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/star.html#star"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.star.star" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs STAR to align .fastq files.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.fastq file</dd>
<dt>output: </dt>
<dd>Aligned.out.bam</dd>
<dt>citation: </dt>
<dd><ol class="first last upperalpha simple">
<li>Dobin et al, Bioinformatics 2012; doi: 10.1093/bioinformatics/bts635 &#8220;STAR: ultrafast universal RNA-seq aligner&#8221;</li>
</ol>
</dd>
<dt>link: </dt>
<dd><a class="reference external" href="https://code.google.com/p/rna-star/">https://code.google.com/p/rna-star/</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">ENDS:</p>
<p>RAW_DATA_DIR:</p>
<p>STAR_INDEX:</p>
<p>STAR_OPTIONS:</p>
<p>STAR_RESULTS:</p>
<p>SAMTOOLS_VERSION:</p>
<p class="last">STAR_VERSION:</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="html-toggle section" id="htseq-count">
<h2>HTSEQ-count<a class="headerlink" href="#htseq-count" title="Permalink to this headline">¶</a></h2>
<span class="target" id="module-omics_pipe.modules.htseq"></span><dl class="function">
<dt id="omics_pipe.modules.htseq.htseq">
<tt class="descclassname">omics_pipe.modules.htseq.</tt><tt class="descname">htseq</tt><big>(</big><em>sample</em>, <em>htseq_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/htseq.html#htseq"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.htseq.htseq" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs htseq-count to get raw count data from alignments.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>Aligned.out.sort.bam</dd>
<dt>output: </dt>
<dd>counts.txt</dd>
<dt>citation: </dt>
<dd>Simon Anders, EMBL</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://www-huber.embl.de/users/anders/HTSeq/doc/overview.html">http://www-huber.embl.de/users/anders/HTSeq/doc/overview.html</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">STAR_RESULTS:</p>
<p>HTSEQ_OPTIONS:</p>
<p>REF_GENES:</p>
<p>HTSEQ_RESULTS:</p>
<p>TEMP_DIR:</p>
<p>SAMTOOLS_VERSION:</p>
<p class="last">BAM_FILE_NAME:</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="html-toggle section" id="module-omics_pipe.modules.rseqc">
<span id="rseqc"></span><h2>RSEQC<a class="headerlink" href="#module-omics_pipe.modules.rseqc" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="omics_pipe.modules.rseqc.rseqc">
<tt class="descclassname">omics_pipe.modules.rseqc.</tt><tt class="descname">rseqc</tt><big>(</big><em>sample</em>, <em>rseqc_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/rseqc.html#rseqc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.rseqc.rseqc" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs rseqc to determine insert size as QC for alignment.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.bam</dd>
<dt>output: </dt>
<dd>pdf plot</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://rseqc.sourceforge.net/">http://rseqc.sourceforge.net/</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">STAR_RESULTS:</p>
<p>QC_PATH:</p>
<p>BAM_FILE_NAME:</p>
<p>RSEQC_REF:</p>
<p>RSEQC_VERSION:</p>
<p class="last">TEMP_DIR:</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="html-toggle section" id="fusion-catcher">
<h2>Fusion Catcher<a class="headerlink" href="#fusion-catcher" title="Permalink to this headline">¶</a></h2>
<span class="target" id="module-omics_pipe.modules.fusion_catcher"></span><dl class="function">
<dt id="omics_pipe.modules.fusion_catcher.fusion_catcher">
<tt class="descclassname">omics_pipe.modules.fusion_catcher.</tt><tt class="descname">fusion_catcher</tt><big>(</big><em>sample</em>, <em>fusion_catcher_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/fusion_catcher.html#fusion_catcher"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.fusion_catcher.fusion_catcher" title="Permalink to this definition">¶</a></dt>
<dd><p>Detects fusion genes in paired-end RNAseq data.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>paired end .fastq files</dd>
<dt>output: </dt>
<dd>list of candidate fusion genes</dd>
<dt>citation: </dt>
<dd><ol class="first last upperalpha simple" start="19">
<li>Kangaspeska, S. Hultsch, H. Edgren, D. Nicorici, A. Murumgi, O.P. Kallioniemi, Reanalysis of RNA-sequencing data reveals several additional fusion genes with multiple isoforms, PLOS One, Oct. 2012. <a class="reference external" href="http://dx.plos.org/10.1371/journal.pone.0048745">http://dx.plos.org/10.1371/journal.pone.0048745</a></li>
</ol>
</dd>
<dt>link: </dt>
<dd><a class="reference external" href="https://code.google.com/p/fusioncatcher">https://code.google.com/p/fusioncatcher</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">ENDS:</p>
<p>RAW_DATA_DIR:</p>
<p>FUSIONCATCHER_RESULTS:</p>
<p>FUSIONCATCHERBUILD_DIR:</p>
<p>TEMP_DIR:</p>
<p>SAMTOOLS_VERSION:</p>
<p>FUSIONCATCHER_VERSION:</p>
<p>FUSIONCATCHER_OPTIONS:</p>
<p class="last">TISSUE:</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="html-toggle section" id="bwa-snpir">
<h2>BWA/SNPiR<a class="headerlink" href="#bwa-snpir" title="Permalink to this headline">¶</a></h2>
<div class="section" id="module-omics_pipe.modules.bwa">
<span id="bwa"></span><h3>BWA<a class="headerlink" href="#module-omics_pipe.modules.bwa" title="Permalink to this headline">¶</a></h3>
<dl class="function">
<dt id="omics_pipe.modules.bwa.bwa1">
<tt class="descclassname">omics_pipe.modules.bwa.</tt><tt class="descname">bwa1</tt><big>(</big><em>sample</em>, <em>bwa1_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/bwa.html#bwa1"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.bwa.bwa1" title="Permalink to this definition">¶</a></dt>
<dd><p>BWA aligner for read1 of paired_end reads.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.fastq</dd>
<dt>output: </dt>
<dd>.sam</dd>
<dt>citation: </dt>
<dd>Li H. and Durbin R. (2009) Fast and accurate short read alignment with Burrows-Wheeler transform. Bioinformatics, 25, 1754-1760. [PMID: 19451168]</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://bio-bwa.sourceforge.net/bwa.shtml">http://bio-bwa.sourceforge.net/bwa.shtml</a></dd>
<dt>parameters from parameters file:</dt>
<dd><p class="first">BWA_RESULTS:</p>
<p>TEMP_DIR:</p>
<p>SAMTOOLS_VERSION:</p>
<p>BWA_VERSION:</p>
<p>BWA_INDEX:</p>
<p>RAW_DATA_DIR:</p>
<p class="last">GATK_READ_GROUP_INFO:</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="omics_pipe.modules.bwa.bwa2">
<tt class="descclassname">omics_pipe.modules.bwa.</tt><tt class="descname">bwa2</tt><big>(</big><em>sample</em>, <em>bwa2_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/bwa.html#bwa2"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.bwa.bwa2" title="Permalink to this definition">¶</a></dt>
<dd><p>BWA aligner for read2 of paired_end reads.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.fastq</dd>
<dt>output: </dt>
<dd>.sam</dd>
<dt>citation: </dt>
<dd>Li H. and Durbin R. (2009) Fast and accurate short read alignment with Burrows-Wheeler transform. Bioinformatics, 25, 1754-1760. [PMID: 19451168]</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://bio-bwa.sourceforge.net/bwa.shtml">http://bio-bwa.sourceforge.net/bwa.shtml</a></dd>
<dt>parameters from parameters file:</dt>
<dd><p class="first">BWA_RESULTS:</p>
<p>TEMP_DIR:</p>
<p>SAMTOOLS_VERSION:</p>
<p>BWA_VERSION:</p>
<p>BWA_INDEX:</p>
<p>RAW_DATA_DIR:</p>
<p class="last">GATK_READ_GROUP_INFO:</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="omics_pipe.modules.bwa.bwa_RNA">
<tt class="descclassname">omics_pipe.modules.bwa.</tt><tt class="descname">bwa_RNA</tt><big>(</big><em>sample</em>, <em>bwa_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/bwa.html#bwa_RNA"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.bwa.bwa_RNA" title="Permalink to this definition">¶</a></dt>
<dd><p>BWA aligner for single end reads.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.fastq</dd>
<dt>output:    </dt>
<dd>.sam</dd>
<dt>citation: </dt>
<dd>Li H. and Durbin R. (2009) Fast and accurate short read alignment with Burrows-Wheeler transform. Bioinformatics, 25, 1754-1760. [PMID: 19451168]</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://bio-bwa.sourceforge.net/bwa.shtml">http://bio-bwa.sourceforge.net/bwa.shtml</a></dd>
<dt>parameters from parameters file:</dt>
<dd><p class="first">BWA_RESULTS:</p>
<p>TEMP_DIR:</p>
<p>SAMTOOLS_VERSION:</p>
<p>BWA_VERSION:</p>
<p>BWA_INDEX:</p>
<p>RAW_DATA_DIR:</p>
<p class="last">GATK_READ_GROUP_INFO:</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-omics_pipe.modules.snpir_variants">
<span id="snpir"></span><h3>SNPiR<a class="headerlink" href="#module-omics_pipe.modules.snpir_variants" title="Permalink to this headline">¶</a></h3>
<dl class="function">
<dt id="omics_pipe.modules.snpir_variants.snpir_variants">
<tt class="descclassname">omics_pipe.modules.snpir_variants.</tt><tt class="descname">snpir_variants</tt><big>(</big><em>sample</em>, <em>snpir_variants_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/snpir_variants.html#snpir_variants"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.snpir_variants.snpir_variants" title="Permalink to this definition">¶</a></dt>
<dd><p>Calls variants using SNPIR pipeline.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>Aligned.out.sort.bam or accepted_hits.bam</dd>
<dt>output: </dt>
<dd>final_variants.vcf file</dd>
<dt>citation: </dt>
<dd>Piskol, R., et al. (2013). &#8220;Reliable Identification of Genomic Variants from RNA-Seq Data.&#8221; The American Journal of Human Genetics 93(4): 641-651.</dd>
<dt>link:</dt>
<dd><a class="reference external" href="http://lilab.stanford.edu/SNPiR/">http://lilab.stanford.edu/SNPiR/</a></dd>
<dt>parameters from parameters file:</dt>
<dd><p class="first">SNPIR_RESULTS:</p>
<p>TEMP_DIR:</p>
<p>SAMTOOLS_VERSION:</p>
<p>BWA_VERSION:</p>
<p>PICARD_VERSION:</p>
<p>GATK_VERSION:</p>
<p>BEDTOOLS_VERSION:</p>
<p>UCSC_TOOLS_VERSION:</p>
<p>GENOME:</p>
<p>REPEAT_MASKER:</p>
<p>SNPIR_ANNOTATION:</p>
<p>RNA_EDIT:</p>
<p>DBSNP:</p>
<p>MILLS:</p>
<p>G1000:</p>
<p>WORKING_DIR:</p>
<p>BWA_RESULTS:</p>
<p>SNPIR_VERSION:</p>
<p>SNPIR_CONFIG:</p>
<p>SNPIR_DIR:</p>
<p class="last">ENCODING:</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-omics_pipe.modules.filter_variants">
<span id="filter-variants"></span><h3>Filter Variants<a class="headerlink" href="#module-omics_pipe.modules.filter_variants" title="Permalink to this headline">¶</a></h3>
<dl class="function">
<dt id="omics_pipe.modules.filter_variants.filter_variants">
<tt class="descclassname">omics_pipe.modules.filter_variants.</tt><tt class="descname">filter_variants</tt><big>(</big><em>sample</em>, <em>filter_variants_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/filter_variants.html#filter_variants"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.filter_variants.filter_variants" title="Permalink to this definition">¶</a></dt>
<dd><p>Filters variants to remove common and low impact variants from the SNPIR pipeline.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.bam or .sam file</dd>
<dt>output: </dt>
<dd>.vcf file</dd>
<dt>citation: </dt>
<dd>Piskol et al. 2013. Reliable identification of genomic variants from RNA-seq data. The American Journal of Human Genetics 93: 641-651.</dd>
<dt>link:</dt>
<dd><a class="reference external" href="http://lilab.stanford.edu/SNPiR/">http://lilab.stanford.edu/SNPiR/</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">SNPIR_RESULTS:</p>
<p>TEMP_DIR:</p>
<p>SAMTOOLS_VERSION:</p>
<p>BWA_VERSION:</p>
<p>PICARD_VERSION:</p>
<p>GATK_VERSION:</p>
<p>BEDTOOLS_VERSION:</p>
<p>UCSC_TOOLS_VERSION:</p>
<p>GENOME:</p>
<p>REPEAT_MASKER:</p>
<p>SNPIR_ANNOTATION:</p>
<p>RNA_EDIT:</p>
<p>DBSNP:</p>
<p>MILLS:</p>
<p>G1000:</p>
<p>WORKING_DIR:</p>
<p>BWA_RESULTS:</p>
<p>SNPIR_VERSION:</p>
<p>SNPIR_CONFIG:</p>
<p>SNPIR_DIR:</p>
<p>SNPEFF_VERSION:</p>
<p>dbNSFP:</p>
<p>VCFTOOLS_VERSION:</p>
<p>WORKING_DIR:</p>
<p class="last">SNP_FILTER_OUT_REF:</p>
</dd>
</dl>
</dd></dl>

</div>
</div>
<div class="html-toggle section" id="module-omics_pipe.modules.intogen">
<span id="intogen"></span><h2>Intogen<a class="headerlink" href="#module-omics_pipe.modules.intogen" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="omics_pipe.modules.intogen.intogen">
<tt class="descclassname">omics_pipe.modules.intogen.</tt><tt class="descname">intogen</tt><big>(</big><em>sample</em>, <em>intogen_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/intogen.html#intogen"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.intogen.intogen" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs Intogen to rank mutations and implication for cancer phenotype. Follows variant calling.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.vcf</dd>
<dt>output: </dt>
<dd>variant list</dd>
<dt>citation: </dt>
<dd>Gonzalez-Perez et al. 2013. Intogen mutations identifies cancer drivers across tumor types. Nature Methods 10, 1081-1082.</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://www.intogen.org/">http://www.intogen.org/</a></dd>
<dt>parameters from parameter file: </dt>
<dd><p class="first">INTOGEN_OPTIONS:</p>
<p>INTOGEN_RESULTS:</p>
<p>INTOGEN_VERSION:</p>
<p>USERNAME:</p>
<p>WORKING_DIR:</p>
<p class="last">TEMP_DIR:</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="html-toggle section" id="oncorep-cancer-report">
<h2>OncoRep Cancer Report<a class="headerlink" href="#oncorep-cancer-report" title="Permalink to this headline">¶</a></h2>
<span class="target" id="module-omics_pipe.modules.BreastCancer_RNA_report"></span><dl class="function">
<dt id="omics_pipe.modules.BreastCancer_RNA_report.BreastCancer_RNA_report">
<tt class="descclassname">omics_pipe.modules.BreastCancer_RNA_report.</tt><tt class="descname">BreastCancer_RNA_report</tt><big>(</big><em>sample</em>, <em>BreastCancer_RNA_report_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/BreastCancer_RNA_report.html#BreastCancer_RNA_report"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.BreastCancer_RNA_report.BreastCancer_RNA_report" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs R script with knitr to produce report from RNAseq pipeline.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>results from other steps in RNAseq pipelines</dd>
<dt>output: </dt>
<dd>html report</dd>
<dt>citation: </dt>
<dd><ol class="first last upperalpha simple" start="20">
<li>Meissner</li>
</ol>
</dd>
<dt>parameters from parameter file:</dt>
<dd><p class="first">WORKING_DIR:</p>
<p>R_VERSION:</p>
<p>REPORT_RESULTS:</p>
<p>PARAMS_FILE:</p>
<p>TABIX_VERSION:</p>
<p>TUMOR_TYPE:</p>
<p>GENELIST:</p>
<p>COSMIC:</p>
<p>CLINVAR:</p>
<p>PHARMGKB_rsID:</p>
<p>PHARMGKB_Allele:</p>
<p>DRUGBANK:</p>
<p class="last">CADD:</p>
</dd>
</dl>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
        <p class="logo"><a href="contents.html" title="contents">
          <img class="logo" src="_static/logo.jpg" alt="Logo"/>
        </a></p><div class="sphinxlocaltoc">
    <h3><a href="contents.html">Page contents</a></h3>
    <ul>
<li><a class="reference internal" href="#">Breast Cancer Personalized Genomics Report- RNAseq</a><ul>
<li><a class="reference internal" href="#module-omics_pipe.modules.fastqc">FASTQC</a></li>
<li><a class="reference internal" href="#star-aligner">STAR Aligner</a></li>
<li><a class="reference internal" href="#htseq-count">HTSEQ-count</a></li>
<li><a class="reference internal" href="#module-omics_pipe.modules.rseqc">RSEQC</a></li>
<li><a class="reference internal" href="#fusion-catcher">Fusion Catcher</a></li>
<li><a class="reference internal" href="#bwa-snpir">BWA/SNPiR</a><ul>
<li><a class="reference internal" href="#module-omics_pipe.modules.bwa">BWA</a></li>
<li><a class="reference internal" href="#module-omics_pipe.modules.snpir_variants">SNPiR</a></li>
<li><a class="reference internal" href="#module-omics_pipe.modules.filter_variants">Filter Variants</a></li>
</ul>
</li>
<li><a class="reference internal" href="#module-omics_pipe.modules.intogen">Intogen</a></li>
<li><a class="reference internal" href="#oncorep-cancer-report">OncoRep Cancer Report</a></li>
</ul>
</li>
</ul>

  </div>
  <div class="sphinxprev">
    <h4>Previous page</h4>
    <p class="topless"><a href="RNAseq_counts_modules.html"
                          title="Previous page">&larr; RNA-seq Count Based Modules</a></p>
  </div>
  <div class="sphinxnext">
    <h4>Next page</h4>
    <p class="topless"><a href="RNAseq_cancer_modules_TCGA.html"
                          title="Next page">&rarr; TCGA Reanalysis Pipeline - RNAseq</a></p>
  </div>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/RNAseq_cancer_modules.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="relbar-bottom">
        
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> &nbsp; &nbsp;</li>
        <li class="right" >
          <a href="RNAseq_cancer_modules_TCGA.html" title="TCGA Reanalysis Pipeline - RNAseq"
             >next</a> &nbsp; &nbsp;</li>
        <li class="right" >
          <a href="RNAseq_counts_modules.html" title="RNA-seq Count Based Modules"
             >previous</a> &nbsp; &nbsp;</li>
    <li><a href="contents.html">Omics Pipe 1.0.16 Documentation</a> &raquo;</li>
 
      </ul>
    </div>
    </div>

    <div class="footer">
        &copy; <a href="copyright.html">Copyright</a> 2014, Kathleen Fisch, Ph.D..
      Last updated on Jul 16, 2014.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.2.
    </div>
    <!-- cloud_sptheme 1.4 -->
  </body>
</html>