QIIME-tools
===========

The qiime-tools project is a collection of python code and scripts that
modifies the original QIIME [1] pipeline by adding or modifying several
steps including support for cluster-computing, multiple primer support
(eliminate primer bias) [2], enhanced support for species-specific
analysis, and additional visualization tools.

Documentation
-------------

Full documentation for the scripts and code is available at the
`qiime-tools documentation site`_

Requirements
------------

`matplotlib`_ for PCoA plots. `Biopython`_ for some sequence and fastq
processing, although its use is being phased out. `fuzzpy`_ for the
otu\_calc module.

Citing
------

A manuscript describing the qiime-tools software is currently in
preparation. Until publication, please cite the github repository and
the author: Shareef M. Dabdoub.

References
----------

[1] J Gregory Caporaso, et al., *QIIME allows analysis of
high-throughput community sequencing data*. Nature Methods, 2010;
`doi:10.1038/nmeth.f.303`_

[2] Kumar PS, et al., *Target Region Selection Is a Critical Determinant
of Community Fingerprints Generated by 16S Pyrosequencing*. PLoS ONE
(2011) 6(6): e20956. `doi:10.1371/journal.pone.0020956`_

.. _qiime-tools documentation site: http://qiime-tools.readthedocs.org
.. _matplotlib: http://matplotlib.org
.. _Biopython: http://biopython.org
.. _fuzzpy: http://mavrinac.com/index.cgi?page=fuzzpy
.. _`doi:10.1038/nmeth.f.303`: http://dx.doi.org/10.1038/nmeth.f.303
.. _`doi:10.1371/journal.pone.0020956`: http://dx.doi.org/10.1371/journal.pone.0020956