Getting Help
------------

There are several ways to get help with the ``hgvs`` package.

The `hgvs-discuss mailing list
<https://groups.google.com/forum/#!forum/hgvs-discuss>`_ is the preferred
way to reach the ``hgvs`` package authors.  Please file bugs and feature
requests on the `hgvs issue tracker
<https://bitbucket.org/invitae/hgvs/issues?status=new&status=open>`_.

If you have questions about the `HGVS Mutation Nomenclature Recommendations
<http://www.hgvs.org/mutnomen/>`_, consider posting your questions to the
`HGVS Facebook page <https://www.facebook.com/HGVSmutnomen>`_.


Links
~~~~~

* `Bioinformatics Data Interface (BDI) <https://bitbucket.org/invitae/bdi/>`_
* `HGVS Mutation Nomenclature Recommendations <http://www.hgvs.org/mutnomen/>`_
* `Human Genome Variation Society (HGVS) <http://www.hgvs.org/>`_
* `Parsley <https://pypi.python.org/pypi/Parsley>`_
* `Universal Transcript Archive (UTA) <https://bitbucket.org/invitae/uta/>`_

References
~~~~~~~~~~
Describing structural changes by extending HGVS sequence variation nomenclature.
  | Taschner, P. E. M., & den Dunnen, J. T.
  | Human mutation, 32(5), 507â€“11. (2011).
  | http://www.ncbi.nlm.nih.gov/pubmed/21309030

A formalized description of the standard human variant nomenclature in Extended Backus-Naur Form.
  | Laros, J. F. J., Blavier, A., den Dunnen, J. T., & Taschner, P. E. M.
  | BMC bioinformatics, 12 Suppl 4(Suppl 4), S5. (2011). 
  | http://www.ncbi.nlm.nih.gov/pubmed/21992071
