


<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>All Available Modules &mdash; Omics Pipe v1.0.16 Documentation</title>
    
    <link rel="stylesheet" href="_static/cloud.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="//fonts.googleapis.com/css?family=Noticia+Text|Open+Sans|Droid+Sans+Mono" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '',
        VERSION:     '1.0.16',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="_static/jquery.cookie.js"></script>
    <script type="text/javascript" src="_static/cloud.js"></script>
    <link rel="author" title="About these documents" href="about.html" />
    <link rel="copyright" title="Copyright" href="copyright.html" />
    <link rel="top" title="Omics Pipe v1.0.16 Documentation" href="index.html" /> 
        <meta name="viewport" content="width=device-width, initial-scale=1">
  </head>
  <body>
    <div class="relbar-top">
        
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> &nbsp; &nbsp;</li>
    <li><a href="contents.html">Omics Pipe 1.0.16 Documentation</a> &raquo;</li>
 
      </ul>
    </div>
    </div>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="all-available-modules">
<h1>All Available Modules<a class="headerlink" href="#all-available-modules" title="Permalink to this headline">¶</a></h1>
<p>Below are all available modules in the current release of Omics Pipe in alphabetical order. When creating
a custom pipeline, you can choose from these modules or create your own.</p>
<span class="target" id="module-omics_pipe.modules.annotate_peaks"></span><dl class="function">
<dt id="omics_pipe.modules.annotate_peaks.annotate_peaks">
<tt class="descclassname">omics_pipe.modules.annotate_peaks.</tt><tt class="descname">annotate_peaks</tt><big>(</big><em>step</em>, <em>annotate_peaks_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/annotate_peaks.html#annotate_peaks"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.annotate_peaks.annotate_peaks" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs HOMER to annotate peak track from ChIPseq data.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.tag input file</dd>
<dt>output: </dt>
<dd>.txt file</dd>
<dt>citation:</dt>
<dd>Heinz S, Benner C, Spann N, Bertolino E et al. Simple Combinations of Lineage-Determining Transcription Factors Prime cis-Regulatory Elements Required for Macrophage and B Cell Identities. Mol Cell 2010 May 28;38(4):576-589. PMID: 20513432</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://homer.salk.edu/homer/">http://homer.salk.edu/homer/</a></dd>
<dt>parameters from parameters file:</dt>
<dd><p class="first">PAIR_LIST:</p>
<p>HOMER_RESULTS:</p>
<p>HOMER_VERSION:</p>
<p>TEMP_DIR:</p>
<p>HOMER_GENOME:</p>
<p class="last">HOMER_ANNOTATE_OPTIONS:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.annotate_variants"></span><dl class="function">
<dt id="omics_pipe.modules.annotate_variants.annotate_variants">
<tt class="descclassname">omics_pipe.modules.annotate_variants.</tt><tt class="descname">annotate_variants</tt><big>(</big><em>sample</em>, <em>annotate_variants_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/annotate_variants.html#annotate_variants"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.annotate_variants.annotate_variants" title="Permalink to this definition">¶</a></dt>
<dd><p>Annotates variants with ANNOVAR. Follows VarCall.
input:</p>
<blockquote>
<div>.vcf</div></blockquote>
<dl class="docutils">
<dt>output: </dt>
<dd>.vcf</dd>
<dt>citation: </dt>
<dd>Wang K, Li M, Hakonarson H. ANNOVAR: Functional annotation of genetic variants from next-generation sequencing data Nucleic Acids Research, 38:e164, 2010</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://www.openbioinformatics.org/annovar/">http://www.openbioinformatics.org/annovar/</a></dd>
<dt>parameters from parameters file:</dt>
<dd><p class="first">VARIANT_RESULTS:</p>
<p>ANNOVARDB:</p>
<p>ANNOVAR_OPTIONS:</p>
<p>ANNOVAR_OPTIONS2:</p>
<p>TEMP_DIR:</p>
<p>ANNOVAR_VERSION:</p>
<p class="last">VCFTOOLS_VERSION:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.bowtie"></span><dl class="function">
<dt id="omics_pipe.modules.bowtie.bowtie">
<tt class="descclassname">omics_pipe.modules.bowtie.</tt><tt class="descname">bowtie</tt><big>(</big><em>sample</em>, <em>bowtie_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/bowtie.html#bowtie"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.bowtie.bowtie" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs Bowtie to align .fastq files.</p>
<dl class="docutils">
<dt>input:</dt>
<dd>.fastq file</dd>
<dt>output:</dt>
<dd>sample.bam</dd>
<dt>citation: </dt>
<dd>Langmead B, Trapnell C, Pop M, Salzberg SL. Ultrafast and memory-efficient alignment of short DNA sequences to the human genome. Genome Biology 10:R25</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://bowtie-bio.sourceforge.net/index.shtml">http://bowtie-bio.sourceforge.net/index.shtml</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">ENDS:</p>
<p>TRIMMED_DATA_PATH:</p>
<p>BOWTIE_OPTIONS:</p>
<p>BOWTIE_INDEX:</p>
<p>BOWTIE_RESULTS:</p>
<p>BOWTIE_VERSION:</p>
<p>SAMTOOLS_VERSION:</p>
<p class="last">BEDTOOLS_VERSION:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.BreastCancer_RNA_report"></span><dl class="function">
<dt id="omics_pipe.modules.BreastCancer_RNA_report.BreastCancer_RNA_report">
<tt class="descclassname">omics_pipe.modules.BreastCancer_RNA_report.</tt><tt class="descname">BreastCancer_RNA_report</tt><big>(</big><em>sample</em>, <em>BreastCancer_RNA_report_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/BreastCancer_RNA_report.html#BreastCancer_RNA_report"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.BreastCancer_RNA_report.BreastCancer_RNA_report" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs R script with knitr to produce report from RNAseq pipeline.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>results from other steps in RNAseq pipelines</dd>
<dt>output: </dt>
<dd>html report</dd>
<dt>citation: </dt>
<dd><ol class="first last upperalpha simple" start="20">
<li>Meissner</li>
</ol>
</dd>
<dt>parameters from parameter file:</dt>
<dd><p class="first">WORKING_DIR:</p>
<p>R_VERSION:</p>
<p>REPORT_RESULTS:</p>
<p>DPS_VERSION:</p>
<p>PARAMS_FILE:</p>
<p>TABIX_VERSION:</p>
<p>TUMOR_TYPE:</p>
<p>GENELIST:</p>
<p>COSMIC:</p>
<p>CLINVAR:</p>
<p>PHARMGKB_rsID:</p>
<p>PHARMGKB_Allele:</p>
<p>DRUGBANK:</p>
<p class="last">CADD:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.bwa"></span><dl class="function">
<dt id="omics_pipe.modules.bwa.bwa1">
<tt class="descclassname">omics_pipe.modules.bwa.</tt><tt class="descname">bwa1</tt><big>(</big><em>sample</em>, <em>bwa1_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/bwa.html#bwa1"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.bwa.bwa1" title="Permalink to this definition">¶</a></dt>
<dd><p>BWA aligner for read1 of paired_end reads.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.fastq</dd>
<dt>output: </dt>
<dd>.sam</dd>
<dt>citation: </dt>
<dd>Li H. and Durbin R. (2009) Fast and accurate short read alignment with Burrows-Wheeler transform. Bioinformatics, 25, 1754-1760. [PMID: 19451168]</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://bio-bwa.sourceforge.net/bwa.shtml">http://bio-bwa.sourceforge.net/bwa.shtml</a></dd>
<dt>parameters from parameters file:</dt>
<dd><p class="first">BWA_RESULTS:</p>
<p>TEMP_DIR:</p>
<p>SAMTOOLS_VERSION:</p>
<p>BWA_VERSION:</p>
<p>BWA_INDEX:</p>
<p>RAW_DATA_DIR:</p>
<p class="last">GATK_READ_GROUP_INFO:</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="omics_pipe.modules.bwa.bwa2">
<tt class="descclassname">omics_pipe.modules.bwa.</tt><tt class="descname">bwa2</tt><big>(</big><em>sample</em>, <em>bwa2_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/bwa.html#bwa2"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.bwa.bwa2" title="Permalink to this definition">¶</a></dt>
<dd><p>BWA aligner for read2 of paired_end reads.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.fastq</dd>
<dt>output: </dt>
<dd>.sam</dd>
<dt>citation: </dt>
<dd>Li H. and Durbin R. (2009) Fast and accurate short read alignment with Burrows-Wheeler transform. Bioinformatics, 25, 1754-1760. [PMID: 19451168]</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://bio-bwa.sourceforge.net/bwa.shtml">http://bio-bwa.sourceforge.net/bwa.shtml</a></dd>
<dt>parameters from parameters file:</dt>
<dd><p class="first">BWA_RESULTS:</p>
<p>TEMP_DIR:</p>
<p>SAMTOOLS_VERSION:</p>
<p>BWA_VERSION:</p>
<p>BWA_INDEX:</p>
<p>RAW_DATA_DIR:</p>
<p class="last">GATK_READ_GROUP_INFO:</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="omics_pipe.modules.bwa.bwa_RNA">
<tt class="descclassname">omics_pipe.modules.bwa.</tt><tt class="descname">bwa_RNA</tt><big>(</big><em>sample</em>, <em>bwa_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/bwa.html#bwa_RNA"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.bwa.bwa_RNA" title="Permalink to this definition">¶</a></dt>
<dd><p>BWA aligner for single end reads.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.fastq</dd>
<dt>output:    </dt>
<dd>.sam</dd>
<dt>citation: </dt>
<dd>Li H. and Durbin R. (2009) Fast and accurate short read alignment with Burrows-Wheeler transform. Bioinformatics, 25, 1754-1760. [PMID: 19451168]</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://bio-bwa.sourceforge.net/bwa.shtml">http://bio-bwa.sourceforge.net/bwa.shtml</a></dd>
<dt>parameters from parameters file:</dt>
<dd><p class="first">BWA_RESULTS:</p>
<p>TEMP_DIR:</p>
<p>SAMTOOLS_VERSION:</p>
<p>BWA_VERSION:</p>
<p>BWA_INDEX:</p>
<p>RAW_DATA_DIR:</p>
<p class="last">GATK_READ_GROUP_INFO:</p>
</dd>
</dl>
</dd></dl>

<dl class="function">
<dt id="omics_pipe.modules.bwa.bwa_mem">
<tt class="descclassname">omics_pipe.modules.bwa.</tt><tt class="descname">bwa_mem</tt><big>(</big><em>sample</em>, <em>bwa_mem_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/bwa.html#bwa_mem"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.bwa.bwa_mem" title="Permalink to this definition">¶</a></dt>
<dd><p>BWA aligner with BWA-MEM algorithm.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.fastq</dd>
<dt>output: </dt>
<dd>.sam</dd>
<dt>citation: </dt>
<dd>Li H. and Durbin R. (2009) Fast and accurate short read alignment with Burrows-Wheeler transform. Bioinformatics, 25, 1754-1760. [PMID: 19451168]</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://bio-bwa.sourceforge.net/bwa.shtml">http://bio-bwa.sourceforge.net/bwa.shtml</a></dd>
<dt>parameters from parameters file:</dt>
<dd><p class="first">BWA_RESULTS:</p>
<p>TEMP_DIR:</p>
<p>SAMTOOLS_VERSION:</p>
<p>BWA_VERSION:</p>
<p>GENOME:</p>
<p>RAW_DATA_DIR:</p>
<p class="last">BWA_OPTIONS:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.call_variants"></span><dl class="function">
<dt id="omics_pipe.modules.call_variants.call_variants">
<tt class="descclassname">omics_pipe.modules.call_variants.</tt><tt class="descname">call_variants</tt><big>(</big><em>sample</em>, <em>call_variants_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/call_variants.html#call_variants"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.call_variants.call_variants" title="Permalink to this definition">¶</a></dt>
<dd><p>Calls variants from alignment .bam files using Varcall.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>Aligned.out.sort.bam or accepted_hits.bam</dd>
<dt>output: </dt>
<dd>.vcf file</dd>
<dt>citation: </dt>
<dd>Erik Aronesty (2011). ea-utils : &#8220;Command-line tools for processing biological sequencing data&#8221;;</dd>
<dt>link: </dt>
<dd><a class="reference external" href="https://code.google.com/p/ea-utils/wiki/Varcall">https://code.google.com/p/ea-utils/wiki/Varcall</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">STAR_RESULTS:</p>
<p>GENOME:</p>
<p>VARSCAN_PATH:</p>
<p>VARSCAN_OPTIONS:</p>
<p>VARIANT_RESULTS:</p>
<p>TEMP_DIR:</p>
<p>SAMTOOLS_VERSION:</p>
<p>ANNOVAR_VERSION:</p>
<p>VCFTOOLS_VERSION:</p>
<p>VARSCAN_VERSION:</p>
<p class="last">SAMTOOLS_OPTIONS:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.ChIP_trim"></span><dl class="function">
<dt id="omics_pipe.modules.ChIP_trim.ChIP_trim">
<tt class="descclassname">omics_pipe.modules.ChIP_trim.</tt><tt class="descname">ChIP_trim</tt><big>(</big><em>sample</em>, <em>ChIP_trim_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/ChIP_trim.html#ChIP_trim"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.ChIP_trim.ChIP_trim" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs Homer Tools to trim adapters from .fastq files.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.fastq file</dd>
<dt>output: </dt>
<dd>.fastq file</dd>
<dt>citation: </dt>
<dd>Heinz S, Benner C, Spann N, Bertolino E et al. Simple Combinations of Lineage-Determining Transcription Factors Prime cis-Regulatory Elements Required for Macrophage and B Cell Identities. Mol Cell 2010 May 28;38(4):576-589. PMID: 20513432</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://homer.salk.edu/homer/">http://homer.salk.edu/homer/</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">ENDS:</p>
<p>RAW_DATA_DIR:</p>
<p>HOMER_TRIM_OPTIONS:</p>
<p>TRIMMED_DATA_PATH:</p>
<p class="last">HOMER_VERSION:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.cuffdiff"></span><dl class="function">
<dt id="omics_pipe.modules.cuffdiff.cuffdiff">
<tt class="descclassname">omics_pipe.modules.cuffdiff.</tt><tt class="descname">cuffdiff</tt><big>(</big><em>step</em>, <em>cuffdiff_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/cuffdiff.html#cuffdiff"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.cuffdiff.cuffdiff" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs Cuffdiff to perform differential expression. Runs after Cufflinks. Part of Tuxedo Suite.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.bam files</dd>
<dt>output: </dt>
<dd>differential expression results</dd>
<dt>citation: </dt>
<dd>Trapnell C, et al. Transcript assembly and quantification by RNA-Seq reveals unannotated transcripts and isoform switching during cell differentiation Nature Biotechnology doi:10.1038/nbt.1621</dd>
<dt>link:</dt>
<dd><a class="reference external" href="http://cufflinks.cbcb.umd.edu/">http://cufflinks.cbcb.umd.edu/</a></dd>
<dt>parameters from parameters file:</dt>
<dd><p class="first">CUFFDIFF_RESULTS:</p>
<p>GENOME:</p>
<p>CUFFDIFF_OPTIONS:</p>
<p>CUFFMERGE_RESULTS:</p>
<p>CUFFDIFF_INPUT_LIST_COND1:</p>
<p>CUFFDIFF_INPUT_LIST_COND2:</p>
<p class="last">CUFFLINKS_VERSION:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.cuffdiff_miRNA"></span><dl class="function">
<dt id="omics_pipe.modules.cuffdiff_miRNA.cuffdiff_miRNA">
<tt class="descclassname">omics_pipe.modules.cuffdiff_miRNA.</tt><tt class="descname">cuffdiff_miRNA</tt><big>(</big><em>step</em>, <em>cuffdiff_miRNA_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/cuffdiff_miRNA.html#cuffdiff_miRNA"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.cuffdiff_miRNA.cuffdiff_miRNA" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs Cuffdiff to perform differential expression. Runs after Cufflinks. Part of Tuxedo Suite.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.bam files</dd>
<dt>output: </dt>
<dd>differential expression results</dd>
<dt>citation: </dt>
<dd>Trapnell C, et al. Transcript assembly and quantification by RNA-Seq reveals unannotated transcripts and isoform switching during cell differentiation Nature Biotechnology doi:10.1038/nbt.1621</dd>
<dt>link:</dt>
<dd><a class="reference external" href="http://cufflinks.cbcb.umd.edu/">http://cufflinks.cbcb.umd.edu/</a></dd>
<dt>parameters from parameters file:</dt>
<dd><p class="first">CUFFDIFF_RESULTS:</p>
<p>GENOME:</p>
<p>CUFFDIFF_OPTIONS:</p>
<p>CUFFMERGE_RESULTS:</p>
<p>CUFFDIFF_INPUT_LIST_COND1:</p>
<p>CUFFDIFF_INPUT_LIST_COND2:</p>
<p class="last">CUFFLINKS_VERSION:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.cufflinks"></span><dl class="function">
<dt id="omics_pipe.modules.cufflinks.cufflinks">
<tt class="descclassname">omics_pipe.modules.cufflinks.</tt><tt class="descname">cufflinks</tt><big>(</big><em>sample</em>, <em>cufflinks_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/cufflinks.html#cufflinks"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.cufflinks.cufflinks" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs cufflinks to assemble .bam files from TopHat.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>accepted_hits.bam</dd>
<dt>output: </dt>
<dd>transcripts.gtf</dd>
<dt>citation: </dt>
<dd>Trapnell C, et al. Transcript assembly and quantification by RNA-Seq reveals unannotated transcripts and isoform switching during cell differentiation  Nature Biotechnology doi:10.1038/nbt.1621</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://cufflinks.cbcb.umd.edu/">http://cufflinks.cbcb.umd.edu/</a></dd>
<dt>parameters from parameters file:</dt>
<dd><p class="first">TOPHAT_RESULTS:</p>
<p>CUFFLINKS_RESULTS:</p>
<p>REF_GENES:</p>
<p>GENOME:</p>
<p>CUFFLINKS_OPTIONS:</p>
<p class="last">CUFFLINKS_VERSION:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.cufflinks_miRNA"></span><dl class="function">
<dt id="omics_pipe.modules.cufflinks_miRNA.cufflinks_miRNA">
<tt class="descclassname">omics_pipe.modules.cufflinks_miRNA.</tt><tt class="descname">cufflinks_miRNA</tt><big>(</big><em>sample</em>, <em>cufflinks_miRNA_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/cufflinks_miRNA.html#cufflinks_miRNA"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.cufflinks_miRNA.cufflinks_miRNA" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs cufflinks to assemble .bam files from TopHat. Takes parameter MIRNA_GTF.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>accepted_hits.bam</dd>
<dt>output: </dt>
<dd>transcripts.gtf</dd>
<dt>citation: </dt>
<dd>Trapnell C, et al. Transcript assembly and quantification by RNA-Seq reveals unannotated transcripts and isoform switching during cell differentiation  Nature Biotechnology doi:10.1038/nbt.1621</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://cufflinks.cbcb.umd.edu/">http://cufflinks.cbcb.umd.edu/</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">TOPHAT_RESULTS:</p>
<p>CUFFLINKS_RESULTS:</p>
<p>miRNA_GTF:</p>
<p>GENOME:</p>
<p>CUFFLINKS_OPTIONS:</p>
<p class="last">CUFFLINKS_VERSION:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.cufflinks_ncRNA"></span><dl class="function">
<dt id="omics_pipe.modules.cufflinks_ncRNA.cufflinks_ncRNA">
<tt class="descclassname">omics_pipe.modules.cufflinks_ncRNA.</tt><tt class="descname">cufflinks_ncRNA</tt><big>(</big><em>sample</em>, <em>cufflinks_ncRNA_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/cufflinks_ncRNA.html#cufflinks_ncRNA"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.cufflinks_ncRNA.cufflinks_ncRNA" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs cufflinks to assemble .bam files from TopHat. Takes parameters LNCRNA_GTF and NONCODE_FASTA.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>accepted_hits.bam</dd>
<dt>output: </dt>
<dd>transcripts.gtf</dd>
<dt>citation: </dt>
<dd>Trapnell C, et al. Transcript assembly and quantification by RNA-Seq reveals unannotated transcripts and isoform switching during cell differentiation  Nature Biotechnology doi:10.1038/nbt.1621</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://cufflinks.cbcb.umd.edu/">http://cufflinks.cbcb.umd.edu/</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">TOPHAT_RESULTS:</p>
<p>CUFFLINKS_RESULTS:</p>
<p>LNCRNA_GTF:</p>
<p>NONCODE_FASTA:</p>
<p>CUFFLINKS_OPTIONS:</p>
<p class="last">CUFFLINKS_VERSION:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.cuffmerge"></span><dl class="function">
<dt id="omics_pipe.modules.cuffmerge.cuffmerge">
<tt class="descclassname">omics_pipe.modules.cuffmerge.</tt><tt class="descname">cuffmerge</tt><big>(</big><em>step</em>, <em>cuffmerge_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/cuffmerge.html#cuffmerge"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.cuffmerge.cuffmerge" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs cuffmerge to merge .gtf files from Cufflinks.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>assembly_GTF_list.txt</dd>
<dt>output: </dt>
<dd>merged.gtf</dd>
<dt>citation: </dt>
<dd>Trapnell C, et al. Transcript assembly and quantification by RNA-Seq reveals unannotated transcripts and isoform switching during cell differentiation  Nature Biotechnology doi:10.1038/nbt.1621</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://cufflinks.cbcb.umd.edu/">http://cufflinks.cbcb.umd.edu/</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">CUFFMERGE_RESULTS:</p>
<p>REF_GENES:</p>
<p>GENOME:</p>
<p>CUFFMERGE_OPTIONS:</p>
<p class="last">CUFFLINKS_VERSION:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.cuffmerge_miRNA"></span><dl class="function">
<dt id="omics_pipe.modules.cuffmerge_miRNA.cuffmerge_miRNA">
<tt class="descclassname">omics_pipe.modules.cuffmerge_miRNA.</tt><tt class="descname">cuffmerge_miRNA</tt><big>(</big><em>step</em>, <em>cuffmerge_miRNA_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/cuffmerge_miRNA.html#cuffmerge_miRNA"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.cuffmerge_miRNA.cuffmerge_miRNA" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs cuffmerge to merge .gtf files from Cufflinks.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>assembly_GTF_list.txt</dd>
<dt>output: </dt>
<dd>merged.gtf</dd>
<dt>citation: </dt>
<dd>Trapnell C, et al. Transcript assembly and quantification by RNA-Seq reveals unannotated transcripts and isoform switching during cell differentiation  Nature Biotechnology doi:10.1038/nbt.1621</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://cufflinks.cbcb.umd.edu/">http://cufflinks.cbcb.umd.edu/</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">CUFFMERGE_RESULTS:</p>
<p>miRNA_GTF:</p>
<p>GENOME:</p>
<p>CUFFMERGE_OPTIONS:</p>
<p class="last">CUFFLINKS_VERSION:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.cuffmergetocompare"></span><dl class="function">
<dt id="omics_pipe.modules.cuffmergetocompare.cuffmergetocompare">
<tt class="descclassname">omics_pipe.modules.cuffmergetocompare.</tt><tt class="descname">cuffmergetocompare</tt><big>(</big><em>step</em>, <em>cuffmergetocompare_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/cuffmergetocompare.html#cuffmergetocompare"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.cuffmergetocompare.cuffmergetocompare" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs cuffcompare to annotate merged .gtf files from Cufflinks.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>assembly_GTF_list.txt</dd>
<dt>output: </dt>
<dd>merged.gtf</dd>
<dt>citation: </dt>
<dd>Trapnell C, et al. Transcript assembly and quantification by RNA-Seq reveals unannotated transcripts and isoform switching during cell differentiation  Nature Biotechnology doi:10.1038/nbt.1621</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://cufflinks.cbcb.umd.edu/">http://cufflinks.cbcb.umd.edu/</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">CUFFMERGE_RESULTS:</p>
<p>REF_GENES:</p>
<p>GENOME:</p>
<p>CUFFMERGETOCOMPARE_OPTIONS:</p>
<p class="last">CUFFLINKS_VERSION:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.cuffmergetocompare_miRNA"></span><dl class="function">
<dt id="omics_pipe.modules.cuffmergetocompare_miRNA.cuffmergetocompare_miRNA">
<tt class="descclassname">omics_pipe.modules.cuffmergetocompare_miRNA.</tt><tt class="descname">cuffmergetocompare_miRNA</tt><big>(</big><em>step</em>, <em>cuffmergetocompare_miRNA_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/cuffmergetocompare_miRNA.html#cuffmergetocompare_miRNA"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.cuffmergetocompare_miRNA.cuffmergetocompare_miRNA" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs cuffcompare to annotate merged .gtf files from Cufflinks.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>assembly_GTF_list.txt</dd>
<dt>output: </dt>
<dd>merged.gtf</dd>
<dt>citation: </dt>
<dd>Trapnell C, et al. Transcript assembly and quantification by RNA-Seq reveals unannotated transcripts and isoform switching during cell differentiation  Nature Biotechnology doi:10.1038/nbt.1621</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://cufflinks.cbcb.umd.edu/">http://cufflinks.cbcb.umd.edu/</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">CUFFMERGE_RESULTS:</p>
<p>miRNA_GTF:</p>
<p>GENOME:</p>
<p>CUFFMERGETOCOMPARE_OPTIONS:</p>
<p class="last">CUFFLINKS_VERSION:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.custom_R_report"></span><dl class="function">
<dt id="omics_pipe.modules.custom_R_report.custom_R_report">
<tt class="descclassname">omics_pipe.modules.custom_R_report.</tt><tt class="descname">custom_R_report</tt><big>(</big><em>sample</em>, <em>custom_R_report_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/custom_R_report.html#custom_R_report"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.custom_R_report.custom_R_report" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs R script with knitr to produce report from omics pipeline.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>results from other steps in RNAseq pipelines</dd>
<dt>output: </dt>
<dd>html report</dd>
<dt>citation: </dt>
<dd><ol class="first last upperalpha simple" start="20">
<li>Meissner</li>
</ol>
</dd>
<dt>parameters from parameter file:</dt>
<dd><p class="first">REPORT_SCRIPT:</p>
<p>R_VERSION:</p>
<p>REPORT_RESULTS:</p>
<p>R_MARKUP_FILE:</p>
<p>DPS_VERSION:</p>
<p class="last">PARAMS_FILE:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.cutadapt_miRNA"></span><dl class="function">
<dt id="omics_pipe.modules.cutadapt_miRNA.cutadapt_miRNA">
<tt class="descclassname">omics_pipe.modules.cutadapt_miRNA.</tt><tt class="descname">cutadapt_miRNA</tt><big>(</big><em>sample</em>, <em>cutadapt_miRNA_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/cutadapt_miRNA.html#cutadapt_miRNA"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.cutadapt_miRNA.cutadapt_miRNA" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs Cutadapt to trim adapters from reads.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.fastq</dd>
<dt>output: </dt>
<dd>.fastq</dd>
<dt>citation: </dt>
<dd>Martin 2011. Cutadapt removes adapter sequences from high-throughput sequencing reads. EMBnet.journal 17: 10-12.</dd>
<dt>link: </dt>
<dd><a class="reference external" href="https://code.google.com/p/cutadapt/">https://code.google.com/p/cutadapt/</a></dd>
<dt>parameters from parameters file: </dt>
<dd>sample, RAW_DATA_DIR, ADAPTER, TRIMMED_DATA_PATH</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.fastq_length_filter_miRNA"></span><dl class="function">
<dt id="omics_pipe.modules.fastq_length_filter_miRNA.fastq_length_filter_miRNA">
<tt class="descclassname">omics_pipe.modules.fastq_length_filter_miRNA.</tt><tt class="descname">fastq_length_filter_miRNA</tt><big>(</big><em>sample</em>, <em>fastq_length_filter_miRNA_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/fastq_length_filter_miRNA.html#fastq_length_filter_miRNA"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.fastq_length_filter_miRNA.fastq_length_filter_miRNA" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs custom Python script to filter miRNA reads by length.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.fastq</dd>
<dt>output: </dt>
<dd>.fastq</dd>
<dt>parameters from parameter file: </dt>
<dd>TRIMMED_DATA_PATH:</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.fastqc"></span><dl class="function">
<dt id="omics_pipe.modules.fastqc.fastqc">
<tt class="descclassname">omics_pipe.modules.fastqc.</tt><tt class="descname">fastqc</tt><big>(</big><em>sample</em>, <em>fastqc_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/fastqc.html#fastqc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.fastqc.fastqc" title="Permalink to this definition">¶</a></dt>
<dd><p>QC check of raw .fastq files using FASTQC.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.fastq file</dd>
<dt>output: </dt>
<dd>folder and zipped folder containing html, txt and image files</dd>
<dt>citation: </dt>
<dd>Babraham Bioinformatics</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">http://www.bioinformatics.babraham.ac.uk/projects/fastqc/</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">RAW_DATA_DIR:</p>
<p>QC_PATH:</p>
<p class="last">FASTQC_VERSION:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.fastqc_miRNA"></span><dl class="function">
<dt id="omics_pipe.modules.fastqc_miRNA.fastqc_miRNA">
<tt class="descclassname">omics_pipe.modules.fastqc_miRNA.</tt><tt class="descname">fastqc_miRNA</tt><big>(</big><em>sample</em>, <em>fastqc_miRNA_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/fastqc_miRNA.html#fastqc_miRNA"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.fastqc_miRNA.fastqc_miRNA" title="Permalink to this definition">¶</a></dt>
<dd><p>QC check of raw .fastq files using FASTQC.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.fastq file</dd>
<dt>output: </dt>
<dd>folder and zipped folder containing html, txt and image files</dd>
<dt>citation: </dt>
<dd>Babraham Bioinformatics</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">http://www.bioinformatics.babraham.ac.uk/projects/fastqc/</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">RAW_DATA_DIR:</p>
<p>QC_PATH:</p>
<p class="last">FASTQC_VERSION:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.filter_variants"></span><dl class="function">
<dt id="omics_pipe.modules.filter_variants.filter_variants">
<tt class="descclassname">omics_pipe.modules.filter_variants.</tt><tt class="descname">filter_variants</tt><big>(</big><em>sample</em>, <em>filter_variants_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/filter_variants.html#filter_variants"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.filter_variants.filter_variants" title="Permalink to this definition">¶</a></dt>
<dd><p>Filters variants to remove common and low impact variants from the SNPIR pipeline.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.bam or .sam file</dd>
<dt>output: </dt>
<dd>.vcf file</dd>
<dt>citation: </dt>
<dd>Piskol et al. 2013. Reliable identification of genomic variants from RNA-seq data. The American Journal of Human Genetics 93: 641-651.</dd>
<dt>link:</dt>
<dd><a class="reference external" href="http://lilab.stanford.edu/SNPiR/">http://lilab.stanford.edu/SNPiR/</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">SNPIR_RESULTS:</p>
<p>TEMP_DIR:</p>
<p>SAMTOOLS_VERSION:</p>
<p>BWA_VERSION:</p>
<p>PICARD_VERSION:</p>
<p>GATK_VERSION:</p>
<p>BEDTOOLS_VERSION:</p>
<p>UCSC_TOOLS_VERSION:</p>
<p>GENOME:</p>
<p>REPEAT_MASKER:</p>
<p>SNPIR_ANNOTATION:</p>
<p>RNA_EDIT:</p>
<p>DBSNP:</p>
<p>MILLS:</p>
<p>G1000:</p>
<p>WORKING_DIR:</p>
<p>BWA_RESULTS:</p>
<p>SNPIR_VERSION:</p>
<p>SNPIR_CONFIG:</p>
<p>SNPIR_DIR:</p>
<p>SNPEFF_VERSION:</p>
<p>dbNSFP:</p>
<p>VCFTOOLS_VERSION:</p>
<p>WORKING_DIR:</p>
<p class="last">SNP_FILTER_OUT_REF:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.find_motifs"></span><dl class="function">
<dt id="omics_pipe.modules.find_motifs.find_motifs">
<tt class="descclassname">omics_pipe.modules.find_motifs.</tt><tt class="descname">find_motifs</tt><big>(</big><em>step</em>, <em>find_motifs_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/find_motifs.html#find_motifs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.find_motifs.find_motifs" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs HOMER to find motifs from ChIPseq data.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.txt peak file from Homer</dd>
<dt>output: </dt>
<dd>.txt file</dd>
<dt>citation: </dt>
<dd>Heinz S, Benner C, Spann N, Bertolino E et al. Simple Combinations of Lineage-Determining Transcription Factors Prime cis-Regulatory Elements Required for Macrophage and B Cell Identities. Mol Cell 2010 May 28;38(4):576-589. PMID: 20513432</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://homer.salk.edu/homer/">http://homer.salk.edu/homer/</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">PAIR_LIST:</p>
<p>HOMER_RESULTS:</p>
<p>HOMER_VERSION:</p>
<p>TEMP_DIR:</p>
<p>HOMER_GENOME:</p>
<p class="last">HOMER_MOTIFS_OPTIONS:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.fusion_catcher"></span><dl class="function">
<dt id="omics_pipe.modules.fusion_catcher.fusion_catcher">
<tt class="descclassname">omics_pipe.modules.fusion_catcher.</tt><tt class="descname">fusion_catcher</tt><big>(</big><em>sample</em>, <em>fusion_catcher_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/fusion_catcher.html#fusion_catcher"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.fusion_catcher.fusion_catcher" title="Permalink to this definition">¶</a></dt>
<dd><p>Detects fusion genes in paired-end RNAseq data.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>paired end .fastq files</dd>
<dt>output: </dt>
<dd>list of candidate fusion genes</dd>
<dt>citation: </dt>
<dd><ol class="first last upperalpha simple" start="19">
<li>Kangaspeska, S. Hultsch, H. Edgren, D. Nicorici, A. Murum�gi, O.P. Kallioniemi, Reanalysis of RNA-sequencing data reveals several additional fusion genes with multiple isoforms, PLOS One, Oct. 2012. <a class="reference external" href="http://dx.plos.org/10.1371/journal.pone.0048745">http://dx.plos.org/10.1371/journal.pone.0048745</a></li>
</ol>
</dd>
<dt>link: </dt>
<dd><a class="reference external" href="https://code.google.com/p/fusioncatcher">https://code.google.com/p/fusioncatcher</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">ENDS:</p>
<p>RAW_DATA_DIR:</p>
<p>FUSIONCATCHER_RESULTS:</p>
<p>FUSIONCATCHERBUILD_DIR:</p>
<p>TEMP_DIR:</p>
<p>SAMTOOLS_VERSION:</p>
<p>FUSIONCATCHER_VERSION:</p>
<p>FUSIONCATCHER_OPTIONS:</p>
<p class="last">TISSUE:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.GATK_preprocessing_WES"></span><dl class="function">
<dt id="omics_pipe.modules.GATK_preprocessing_WES.GATK_preprocessing_WES">
<tt class="descclassname">omics_pipe.modules.GATK_preprocessing_WES.</tt><tt class="descname">GATK_preprocessing_WES</tt><big>(</big><em>sample</em>, <em>GATK_preprocessing_WES_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/GATK_preprocessing_WES.html#GATK_preprocessing_WES"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.GATK_preprocessing_WES.GATK_preprocessing_WES" title="Permalink to this definition">¶</a></dt>
<dd><p>GATK preprocessing steps for whole exome sequencing.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>sorted.rg.md.bam</dd>
<dt>output: </dt>
<dd>.ready.bam</dd>
<dt>citation: </dt>
<dd>McKenna A, Hanna M, Banks E, Sivachenko A, Cibulskis K, Kernytsky A, Garimella K, Altshuler D, Gabriel S, Daly M, DePristo MA (2010). The Genome Analysis Toolkit: a MapReduce framework for analyzing next-generation DNA sequencing data. Genome Res. 20:1297-303.</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://www.broadinstitute.org/gatk/">http://www.broadinstitute.org/gatk/</a></dd>
<dt>parameters from parameters file:</dt>
<dd><p class="first">BWA_RESULTS:</p>
<p>TEMP_DIR:</p>
<p>GATK_VERSION:</p>
<p>GENOME:</p>
<p>DBSNP:</p>
<p>MILLS:</p>
<p>G1000:</p>
<p class="last">CAPTURE_KIT_BED:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.GATK_preprocessing_WGS"></span><dl class="function">
<dt id="omics_pipe.modules.GATK_preprocessing_WGS.GATK_preprocessing_WGS">
<tt class="descclassname">omics_pipe.modules.GATK_preprocessing_WGS.</tt><tt class="descname">GATK_preprocessing_WGS</tt><big>(</big><em>sample</em>, <em>GATK_preprocessing_WGS_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/GATK_preprocessing_WGS.html#GATK_preprocessing_WGS"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.GATK_preprocessing_WGS.GATK_preprocessing_WGS" title="Permalink to this definition">¶</a></dt>
<dd><p>GATK preprocessing steps for whole genome sequencing.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>sorted.rg.md.bam</dd>
<dt>output: </dt>
<dd>.ready.bam</dd>
<dt>citation: </dt>
<dd>McKenna A, Hanna M, Banks E, Sivachenko A, Cibulskis K, Kernytsky A, Garimella K, Altshuler D, Gabriel S, Daly M, DePristo MA (2010). The Genome Analysis Toolkit: a MapReduce framework for analyzing next-generation DNA sequencing data. Genome Res. 20:1297-303.</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://www.broadinstitute.org/gatk/">http://www.broadinstitute.org/gatk/</a></dd>
<dt>parameters from parameters file:</dt>
<dd><p class="first">BWA_RESULTS:</p>
<p>TEMP_DIR:</p>
<p>GATK_VERSION:</p>
<p>GENOME:</p>
<p>DBSNP:</p>
<p>MILLS:</p>
<p class="last">G1000:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.GATK_variant_discovery"></span><dl class="function">
<dt id="omics_pipe.modules.GATK_variant_discovery.GATK_variant_discovery">
<tt class="descclassname">omics_pipe.modules.GATK_variant_discovery.</tt><tt class="descname">GATK_variant_discovery</tt><big>(</big><em>sample</em>, <em>GATK_variant_discovery_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/GATK_variant_discovery.html#GATK_variant_discovery"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.GATK_variant_discovery.GATK_variant_discovery" title="Permalink to this definition">¶</a></dt>
<dd><p>GATK_variant_discovery.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>sorted.rg.md.bam</dd>
<dt>output: </dt>
<dd>.ready.bam</dd>
<dt>citation: </dt>
<dd>McKenna A, Hanna M, Banks E, Sivachenko A, Cibulskis K, Kernytsky A, Garimella K, Altshuler D, Gabriel S, Daly M, DePristo MA (2010). The Genome Analysis Toolkit: a MapReduce framework for analyzing next-generation DNA sequencing data. Genome Res. 20:1297-303.</dd>
<dt>link: GATK_variant_discovery</dt>
<dd><a class="reference external" href="http://www.broadinstitute.org/gatk/">http://www.broadinstitute.org/gatk/</a></dd>
<dt>parameters from parameters file:</dt>
<dd><p class="first">BWA_RESULTS:</p>
<p>TEMP_DIR:</p>
<p>GATK_VERSION:</p>
<p>GENOME:</p>
<p class="last">DBSNP:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.GATK_variant_filtering"></span><dl class="function">
<dt id="omics_pipe.modules.GATK_variant_filtering.GATK_variant_filtering">
<tt class="descclassname">omics_pipe.modules.GATK_variant_filtering.</tt><tt class="descname">GATK_variant_filtering</tt><big>(</big><em>sample</em>, <em>GATK_variant_filtering_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/GATK_variant_filtering.html#GATK_variant_filtering"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.GATK_variant_filtering.GATK_variant_filtering" title="Permalink to this definition">¶</a></dt>
<dd><p>GATK_variant_filtering.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>sorted.rg.md.bam</dd>
<dt>output: </dt>
<dd>.ready.bam</dd>
<dt>citation: </dt>
<dd>McKenna A, Hanna M, Banks E, Sivachenko A, Cibulskis K, Kernytsky A, Garimella K, Altshuler D, Gabriel S, Daly M, DePristo MA (2010). The Genome Analysis Toolkit: a MapReduce framework for analyzing next-generation DNA sequencing data. Genome Res. 20:1297-303.</dd>
<dt>link: GATK_variant_filtering</dt>
<dd><a class="reference external" href="http://www.broadinstitute.org/gatk/">http://www.broadinstitute.org/gatk/</a></dd>
<dt>parameters from parameters file:</dt>
<dd><p class="first">BWA_RESULTS:</p>
<p>TEMP_DIR:</p>
<p>GATK_VERSION:</p>
<p>GENOME:</p>
<p>DBSNP:</p>
<p>MILLS_G1000:</p>
<p>OMNI:</p>
<p class="last">HAPMAP:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.homer_peaks"></span><dl class="function">
<dt id="omics_pipe.modules.homer_peaks.homer_peaks">
<tt class="descclassname">omics_pipe.modules.homer_peaks.</tt><tt class="descname">homer_peaks</tt><big>(</big><em>step</em>, <em>homer_peaks_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/homer_peaks.html#homer_peaks"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.homer_peaks.homer_peaks" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs HOMER to call peaks from ChIPseq data.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.tag input file</dd>
<dt>output: </dt>
<dd>.txt file</dd>
<dt>citation: </dt>
<dd>Heinz S, Benner C, Spann N, Bertolino E et al. Simple Combinations of Lineage-Determining Transcription Factors Prime cis-Regulatory Elements Required for Macrophage and B Cell Identities. Mol Cell 2010 May 28;38(4):576-589. PMID: 20513432</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://homer.salk.edu/homer/">http://homer.salk.edu/homer/</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">PAIR_LIST:</p>
<p>HOMER_RESULTS:</p>
<p>HOMER_PEAKS_OPTIONS:</p>
<p>HOMER_VERSION:</p>
<p class="last">TEMP_DIR:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.htseq"></span><dl class="function">
<dt id="omics_pipe.modules.htseq.htseq">
<tt class="descclassname">omics_pipe.modules.htseq.</tt><tt class="descname">htseq</tt><big>(</big><em>sample</em>, <em>htseq_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/htseq.html#htseq"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.htseq.htseq" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs htseq-count to get raw count data from alignments.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>Aligned.out.sort.bam</dd>
<dt>output: </dt>
<dd>counts.txt</dd>
<dt>citation: </dt>
<dd>Simon Anders, EMBL</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://www-huber.embl.de/users/anders/HTSeq/doc/overview.html">http://www-huber.embl.de/users/anders/HTSeq/doc/overview.html</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">STAR_RESULTS:</p>
<p>HTSEQ_OPTIONS:</p>
<p>REF_GENES:</p>
<p>HTSEQ_RESULTS:</p>
<p>TEMP_DIR:</p>
<p>SAMTOOLS_VERSION:</p>
<p class="last">BAM_FILE_NAME:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.htseq_gencode"></span><dl class="function">
<dt id="omics_pipe.modules.htseq_gencode.htseq_gencode">
<tt class="descclassname">omics_pipe.modules.htseq_gencode.</tt><tt class="descname">htseq_gencode</tt><big>(</big><em>sample</em>, <em>htseq_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/htseq_gencode.html#htseq_gencode"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.htseq_gencode.htseq_gencode" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs htseq-count to get raw count data from alignments.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>Aligned.out.sort.bam</dd>
<dt>output: </dt>
<dd>counts.txt</dd>
<dt>citation: </dt>
<dd>Simon Anders, EMBL</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://www-huber.embl.de/users/anders/HTSeq/doc/overview.html">http://www-huber.embl.de/users/anders/HTSeq/doc/overview.html</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">STAR_RESULTS:</p>
<p>HTSEQ_OPTIONS:</p>
<p>REF_GENES_GENCODE:</p>
<p>HTSEQ_GENCODE_RESULTS:</p>
<p>TEMP_DIR:</p>
<p>SAMTOOLS_VERSION:</p>
<p class="last">BAM_FILE_NAME:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.htseq_miRNA"></span><dl class="function">
<dt id="omics_pipe.modules.htseq_miRNA.htseq_miRNA">
<tt class="descclassname">omics_pipe.modules.htseq_miRNA.</tt><tt class="descname">htseq_miRNA</tt><big>(</big><em>sample</em>, <em>htseq_miRNA_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/htseq_miRNA.html#htseq_miRNA"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.htseq_miRNA.htseq_miRNA" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs htseq-count to get raw count data from alignments.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>Aligned.out.sort.bam</dd>
<dt>output: </dt>
<dd>counts.txt</dd>
<dt>citation: </dt>
<dd>Simon Anders, EMBL</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://www-huber.embl.de/users/anders/HTSeq/doc/overview.html">http://www-huber.embl.de/users/anders/HTSeq/doc/overview.html</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">TOPHAT_RESULTS:</p>
<p>HTSEQ_OPTIONS:</p>
<p>miRNA_GFF:</p>
<p>HTSEQ_RESULTS:</p>
<p>TEMP_DIR:</p>
<p>SAMTOOLS_VERSION:</p>
<p class="last">BAM_FILE_NAME:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.intogen"></span><dl class="function">
<dt id="omics_pipe.modules.intogen.intogen">
<tt class="descclassname">omics_pipe.modules.intogen.</tt><tt class="descname">intogen</tt><big>(</big><em>sample</em>, <em>intogen_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/intogen.html#intogen"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.intogen.intogen" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs Intogen to rank mutations and implication for cancer phenotype. Follows variant calling.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.vcf</dd>
<dt>output: </dt>
<dd>variant list</dd>
<dt>citation: </dt>
<dd>Gonzalez-Perez et al. 2013. Intogen mutations identifies cancer drivers across tumor types. Nature Methods 10, 1081-1082.</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://www.intogen.org/">http://www.intogen.org/</a></dd>
<dt>parameters from parameter file: </dt>
<dd><p class="first">INTOGEN_OPTIONS:</p>
<p>INTOGEN_RESULTS:</p>
<p>INTOGEN_VERSION:</p>
<p>USERNAME:</p>
<p>WORKING_DIR:</p>
<p class="last">TEMP_DIR:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.macs"></span><dl class="function">
<dt id="omics_pipe.modules.macs.macs">
<tt class="descclassname">omics_pipe.modules.macs.</tt><tt class="descname">macs</tt><big>(</big><em>step</em>, <em>macs_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/macs.html#macs"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.macs.macs" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs MACS to call peaks from ChIPseq data.
input:</p>
<blockquote>
<div>.fastq file</div></blockquote>
<dl class="docutils">
<dt>output: </dt>
<dd>.fastq file</dd>
<dt>citation: </dt>
<dd>Zhang et al. Model-based Analysis of ChIP-Seq (MACS). Genome Biol (2008) vol. 9 (9) pp. R137</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://liulab.dfci.harvard.edu/MACS/">http://liulab.dfci.harvard.edu/MACS/</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">PAIR_LIST:</p>
<p>BOWTIE_RESULTS:</p>
<p>CHROM_SIZES:</p>
<p>MACS_RESULTS:</p>
<p>MACS_VERSION:</p>
<p>TEMP_DIR:</p>
<p class="last">BEDTOOLS_VERSION:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.mutect"></span><dl class="function">
<dt id="omics_pipe.modules.mutect.mutect">
<tt class="descclassname">omics_pipe.modules.mutect.</tt><tt class="descname">mutect</tt><big>(</big><em>sample</em>, <em>mutect_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/mutect.html#mutect"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.mutect.mutect" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs MuTect on paired tumor/normal samples to detect somatic point mutations in cancer genomes.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.bam</dd>
<dt>output: </dt>
<dd>call_stats.txt</dd>
<dt>citation: </dt>
<dd>Cibulskis, K. et al. Sensitive detection of somatic point mutations in impure and heterogeneous cancer samples. Nat Biotechnology (2013).doi:10.1038/nbt.2514</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://www.broadinstitute.org/cancer/cga/mutect">http://www.broadinstitute.org/cancer/cga/mutect</a></dd>
<dt>parameters from parameters file:</dt>
<dd><p class="first">BWA_RESULTS:</p>
<p>TEMP_DIR:</p>
<p>GATK_VERSION:</p>
<p>GENOME:</p>
<p>DBSNP:</p>
<p>MILLS:</p>
<p>G1000:</p>
<p class="last">CAPTURE_KIT_BED:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.peak_track"></span><dl class="function">
<dt id="omics_pipe.modules.peak_track.peak_track">
<tt class="descclassname">omics_pipe.modules.peak_track.</tt><tt class="descname">peak_track</tt><big>(</big><em>step</em>, <em>peak_track_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/peak_track.html#peak_track"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.peak_track.peak_track" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs HOMER to create peak track from ChIPseq data.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.tag input file</dd>
<dt>output: </dt>
<dd>.txt file</dd>
<dt>citation: </dt>
<dd>Heinz S, Benner C, Spann N, Bertolino E et al. Simple Combinations of Lineage-Determining Transcription Factors Prime cis-Regulatory Elements Required for Macrophage and B Cell Identities. Mol Cell 2010 May 28;38(4):576-589. PMID: 20513432</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://homer.salk.edu/homer/">http://homer.salk.edu/homer/</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">PAIR_LIST:</p>
<p>HOMER_RESULTS:</p>
<p>HOMER_VERSION:</p>
<p class="last">TEMP_DIR:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.picard_mark_duplicates"></span><dl class="function">
<dt id="omics_pipe.modules.picard_mark_duplicates.picard_mark_duplicates">
<tt class="descclassname">omics_pipe.modules.picard_mark_duplicates.</tt><tt class="descname">picard_mark_duplicates</tt><big>(</big><em>sample</em>, <em>picard_mark_duplicates_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/picard_mark_duplicates.html#picard_mark_duplicates"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.picard_mark_duplicates.picard_mark_duplicates" title="Permalink to this definition">¶</a></dt>
<dd><p>Picard tools Mark Duplicates.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>sorted.bam</dd>
<dt>output: </dt>
<dd>_sorted.rg.md.bam</dd>
<dt>citation: </dt>
<dd><a class="reference external" href="http://picard.sourceforge.net/">http://picard.sourceforge.net/</a></dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://picard.sourceforge.net/">http://picard.sourceforge.net/</a></dd>
<dt>parameters from parameters file:</dt>
<dd><p class="first">BWA_RESULTS:</p>
<p>TEMP_DIR:</p>
<p>PICARD_VERSION:</p>
<p class="last">SAMTOOLS_VERSION:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.read_density"></span><dl class="function">
<dt id="omics_pipe.modules.read_density.read_density">
<tt class="descclassname">omics_pipe.modules.read_density.</tt><tt class="descname">read_density</tt><big>(</big><em>sample</em>, <em>read_density_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/read_density.html#read_density"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.read_density.read_density" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs HOMER to visualize read density from ChIPseq data.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.bam file</dd>
<dt>output: </dt>
<dd>.txt file</dd>
<dt>citation: </dt>
<dd>Heinz S, Benner C, Spann N, Bertolino E et al. Simple Combinations of Lineage-Determining Transcription Factors Prime cis-Regulatory Elements Required for Macrophage and B Cell Identities. Mol Cell 2010 May 28;38(4):576-589. PMID: 20513432</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://homer.salk.edu/homer/">http://homer.salk.edu/homer/</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">BOWTIE_RESULTS:</p>
<p>CHROM_SIZES:</p>
<p>HOMER_RESULTS:</p>
<p>HOMER_VERSION:</p>
<p class="last">TEMP_DIR:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.RNAseq_QC"></span><dl class="function">
<dt id="omics_pipe.modules.RNAseq_QC.RNAseq_QC">
<tt class="descclassname">omics_pipe.modules.RNAseq_QC.</tt><tt class="descname">RNAseq_QC</tt><big>(</big><em>sample</em>, <em>RNAseq_QC_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/RNAseq_QC.html#RNAseq_QC"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.RNAseq_QC.RNAseq_QC" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs rseqc to determine insert size as QC for alignment.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.bam</dd>
<dt>output: </dt>
<dd>pdf plot</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://rseqc.sourceforge.net/">http://rseqc.sourceforge.net/</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">STAR_RESULTS:</p>
<p>QC_PATH:</p>
<p>BAM_FILE_NAME:</p>
<p>RSEQC_REF:</p>
<p>RSEQC_VERSION:</p>
<p>TEMP_DIR:</p>
<p class="last">PICARD_VERSION:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.RNAseq_report"></span><dl class="function">
<dt id="omics_pipe.modules.RNAseq_report.RNAseq_report">
<tt class="descclassname">omics_pipe.modules.RNAseq_report.</tt><tt class="descname">RNAseq_report</tt><big>(</big><em>sample</em>, <em>RNAseq_report_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/RNAseq_report.html#RNAseq_report"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.RNAseq_report.RNAseq_report" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs R script with knitr to produce report from RNAseq pipeline.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>results from other steps in RNAseq pipelines</dd>
<dt>output: </dt>
<dd>html report</dd>
<dt>citation: </dt>
<dd><ol class="first last upperalpha simple" start="20">
<li>Meissner</li>
</ol>
</dd>
<dt>parameters from parameter file:</dt>
<dd><p class="first">REPORT_SCRIPT:</p>
<p>R_VERSION:</p>
<p>REPORT_RESULTS:</p>
<p>R_MARKUP_FILE:</p>
<p>DPS_VERSION:</p>
<p class="last">PARAMS_FILE:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.RNAseq_report_counts"></span><dl class="function">
<dt id="omics_pipe.modules.RNAseq_report_counts.RNAseq_report_counts">
<tt class="descclassname">omics_pipe.modules.RNAseq_report_counts.</tt><tt class="descname">RNAseq_report_counts</tt><big>(</big><em>sample</em>, <em>RNAseq_report_counts_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/RNAseq_report_counts.html#RNAseq_report_counts"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.RNAseq_report_counts.RNAseq_report_counts" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs R script with knitr to produce report from RNAseq pipeline.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>results from other steps in RNAseq pipelines</dd>
<dt>output: </dt>
<dd>html report</dd>
<dt>citation: </dt>
<dd><ol class="first last upperalpha simple" start="20">
<li>Meissner</li>
</ol>
</dd>
<dt>parameters from parameter file:</dt>
<dd><p class="first">WORKING_DIR:</p>
<p>R_VERSION:</p>
<p>REPORT_RESULTS:</p>
<p>DPS_VERSION:</p>
<p class="last">PARAMS_FILE:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.rseqc"></span><dl class="function">
<dt id="omics_pipe.modules.rseqc.rseqc">
<tt class="descclassname">omics_pipe.modules.rseqc.</tt><tt class="descname">rseqc</tt><big>(</big><em>sample</em>, <em>rseqc_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/rseqc.html#rseqc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.rseqc.rseqc" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs rseqc to determine insert size as QC for alignment.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.bam</dd>
<dt>output: </dt>
<dd>pdf plot</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://rseqc.sourceforge.net/">http://rseqc.sourceforge.net/</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">STAR_RESULTS:</p>
<p>QC_PATH:</p>
<p>BAM_FILE_NAME:</p>
<p>RSEQC_REF:</p>
<p>RSEQC_VERSION:</p>
<p class="last">TEMP_DIR:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.snpir_variants"></span><dl class="function">
<dt id="omics_pipe.modules.snpir_variants.snpir_variants">
<tt class="descclassname">omics_pipe.modules.snpir_variants.</tt><tt class="descname">snpir_variants</tt><big>(</big><em>sample</em>, <em>snpir_variants_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/snpir_variants.html#snpir_variants"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.snpir_variants.snpir_variants" title="Permalink to this definition">¶</a></dt>
<dd><p>Calls variants using SNPIR pipeline.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>Aligned.out.sort.bam or accepted_hits.bam</dd>
<dt>output: </dt>
<dd>final_variants.vcf file</dd>
<dt>citation: </dt>
<dd>Piskol, R., et al. (2013). &#8220;Reliable Identification of Genomic Variants from RNA-Seq Data.&#8221; The American Journal of Human Genetics 93(4): 641-651.</dd>
<dt>link:</dt>
<dd><a class="reference external" href="http://lilab.stanford.edu/SNPiR/">http://lilab.stanford.edu/SNPiR/</a></dd>
<dt>parameters from parameters file:</dt>
<dd><p class="first">SNPIR_RESULTS:</p>
<p>TEMP_DIR:</p>
<p>SAMTOOLS_VERSION:</p>
<p>BWA_VERSION:</p>
<p>PICARD_VERSION:</p>
<p>GATK_VERSION:</p>
<p>BEDTOOLS_VERSION:</p>
<p>UCSC_TOOLS_VERSION:</p>
<p>GENOME:</p>
<p>REPEAT_MASKER:</p>
<p>SNPIR_ANNOTATION:</p>
<p>RNA_EDIT:</p>
<p>DBSNP:</p>
<p>MILLS:</p>
<p>G1000:</p>
<p>WORKING_DIR:</p>
<p>BWA_RESULTS:</p>
<p>SNPIR_VERSION:</p>
<p>SNPIR_CONFIG:</p>
<p>SNPIR_DIR:</p>
<p class="last">ENCODING:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.star"></span><dl class="function">
<dt id="omics_pipe.modules.star.star">
<tt class="descclassname">omics_pipe.modules.star.</tt><tt class="descname">star</tt><big>(</big><em>sample</em>, <em>star_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/star.html#star"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.star.star" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs STAR to align .fastq files.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.fastq file</dd>
<dt>output: </dt>
<dd>Aligned.out.bam</dd>
<dt>citation: </dt>
<dd><ol class="first last upperalpha simple">
<li>Dobin et al, Bioinformatics 2012; doi: 10.1093/bioinformatics/bts635 &#8220;STAR: ultrafast universal RNA-seq aligner&#8221;</li>
</ol>
</dd>
<dt>link: </dt>
<dd><a class="reference external" href="https://code.google.com/p/rna-star/">https://code.google.com/p/rna-star/</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">ENDS:</p>
<p>RAW_DATA_DIR:</p>
<p>STAR_INDEX:</p>
<p>STAR_OPTIONS:</p>
<p>STAR_RESULTS:</p>
<p>SAMTOOLS_VERSION:</p>
<p class="last">STAR_VERSION:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.star_piRNA"></span><dl class="function">
<dt id="omics_pipe.modules.star_piRNA.star_piRNA">
<tt class="descclassname">omics_pipe.modules.star_piRNA.</tt><tt class="descname">star_piRNA</tt><big>(</big><em>sample</em>, <em>star_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/star_piRNA.html#star_piRNA"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.star_piRNA.star_piRNA" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs STAR to align .fastq files.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.fastq file</dd>
<dt>output: </dt>
<dd>Aligned.out.bam</dd>
<dt>citation: </dt>
<dd><ol class="first last upperalpha simple">
<li>Dobin et al, Bioinformatics 2012; doi: 10.1093/bioinformatics/bts635 &#8220;STAR: ultrafast universal RNA-seq aligner&#8221;</li>
</ol>
</dd>
<dt>link: </dt>
<dd><a class="reference external" href="https://code.google.com/p/rna-star/">https://code.google.com/p/rna-star/</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">ENDS:</p>
<p>RAW_DATA_DIR:</p>
<p>STAR_INDEX:</p>
<p>STAR_OPTIONS:</p>
<p>STAR_RESULTS:</p>
<p>SAMTOOLS_VERSION:</p>
<p class="last">STAR_VERSION:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.TCGA_download"></span><dl class="function">
<dt id="omics_pipe.modules.TCGA_download.TCGA_download">
<tt class="descclassname">omics_pipe.modules.TCGA_download.</tt><tt class="descname">TCGA_download</tt><big>(</big><em>sample</em>, <em>TCGA_download_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/TCGA_download.html#TCGA_download"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.TCGA_download.TCGA_download" title="Permalink to this definition">¶</a></dt>
<dd><p>Downloads and unzips TCGA data from xml file.
input:</p>
<blockquote>
<div>TGCA XML file</div></blockquote>
<dl class="docutils">
<dt>output:</dt>
<dd>downloaded files from TCGA</dd>
<dt>citation: </dt>
<dd>The Cancer Genome Atlas</dd>
<dt>link: </dt>
<dd><a class="reference external" href="https://tcga-data.nci.nih.gov/tcga/tcgaHome2.jsp">https://tcga-data.nci.nih.gov/tcga/tcgaHome2.jsp</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">TCGA_XML_FILE:</p>
<p>TCGA_KEY:</p>
<p>TCGA_OUTPUT_PATH:</p>
<p>SSH_USER_NAME:</p>
<p class="last">CGATOOLS_VERSION:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.tophat"></span><dl class="function">
<dt id="omics_pipe.modules.tophat.tophat">
<tt class="descclassname">omics_pipe.modules.tophat.</tt><tt class="descname">tophat</tt><big>(</big><em>sample</em>, <em>tophat_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/tophat.html#tophat"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.tophat.tophat" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs TopHat to align .fastq files.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.fastq file</dd>
<dt>output: </dt>
<dd>accepted_hits.bam</dd>
<dt>citation: </dt>
<dd>Trapnell C, Pachter L, Salzberg SL. TopHat: discovering splice junctions with RNA-Seq. Bioinformatics doi:10.1093/bioinformatics/btp120</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://tophat.cbcb.umd.edu/">http://tophat.cbcb.umd.edu/</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">RAW_DATA_DIR:</p>
<p>REF_GENES:</p>
<p>TOPHAT_RESULTS:</p>
<p>BOWTIE_INDEX:</p>
<p>TOPHAT_VERSION:</p>
<p class="last">TOPHAT_OPTIONS:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.tophat_miRNA"></span><dl class="function">
<dt id="omics_pipe.modules.tophat_miRNA.tophat_miRNA">
<tt class="descclassname">omics_pipe.modules.tophat_miRNA.</tt><tt class="descname">tophat_miRNA</tt><big>(</big><em>sample</em>, <em>tophat_miRNA_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/tophat_miRNA.html#tophat_miRNA"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.tophat_miRNA.tophat_miRNA" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs TopHat to align .fastq files.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.fastq file</dd>
<dt>output: </dt>
<dd>accepted_hits.bam</dd>
<dt>citation: </dt>
<dd>Trapnell C, Pachter L, Salzberg SL. TopHat: discovering splice junctions with RNA-Seq. Bioinformatics doi:10.1093/bioinformatics/btp120</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://tophat.cbcb.umd.edu/">http://tophat.cbcb.umd.edu/</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">RAW_DATA_DIR:</p>
<p>REF_GENES:</p>
<p>TOPHAT_RESULTS:</p>
<p>miRNA_BOWTIE_INDEX:</p>
<p>TOPHAT_VERSION:</p>
<p class="last">TOPHAT_OPTIONS:</p>
</dd>
</dl>
</dd></dl>

<span class="target" id="module-omics_pipe.modules.tophat_ncRNA"></span><dl class="function">
<dt id="omics_pipe.modules.tophat_ncRNA.tophat_ncRNA">
<tt class="descclassname">omics_pipe.modules.tophat_ncRNA.</tt><tt class="descname">tophat_ncRNA</tt><big>(</big><em>sample</em>, <em>tophat_ncRNA_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/tophat_ncRNA.html#tophat_ncRNA"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.tophat_ncRNA.tophat_ncRNA" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs TopHat to align .fastq files.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.fastq file</dd>
<dt>output: </dt>
<dd>accepted_hits.bam</dd>
<dt>citation: </dt>
<dd>Trapnell C, Pachter L, Salzberg SL. TopHat: discovering splice junctions with RNA-Seq. Bioinformatics doi:10.1093/bioinformatics/btp120</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://tophat.cbcb.umd.edu/">http://tophat.cbcb.umd.edu/</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">RAW_DATA_DIR:</p>
<p>REF_GENES:</p>
<p>TOPHAT_RESULTS:</p>
<p>NONCODE_BOWTIE_INDEX:</p>
<p>TOPHAT_VERSION:</p>
<p class="last">TOPHAT_OPTIONS:</p>
</dd>
</dl>
</dd></dl>

</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
        <p class="logo"><a href="contents.html" title="contents">
          <img class="logo" src="_static/logo.jpg" alt="Logo"/>
        </a></p>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/all_modules.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="relbar-bottom">
        
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> &nbsp; &nbsp;</li>
    <li><a href="contents.html">Omics Pipe 1.0.16 Documentation</a> &raquo;</li>
 
      </ul>
    </div>
    </div>

    <div class="footer">
        &copy; <a href="copyright.html">Copyright</a> 2014, Kathleen Fisch, Ph.D..
      Last updated on Jul 09, 2014.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.1.3.
    </div>
    <!-- cloud_sptheme 1.4 -->
  </body>
</html>