


<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>miRNA-seq Count Based Modules &mdash; Omics Pipe v1.0.16 Documentation</title>
    
    <link rel="stylesheet" href="_static/cloud.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="//fonts.googleapis.com/css?family=Noticia+Text|Open+Sans|Droid+Sans+Mono" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '',
        VERSION:     '1.0.16',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="http://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <script type="text/javascript" src="_static/jquery.cookie.js"></script>
    <script type="text/javascript" src="_static/cloud.js"></script>
    <link rel="author" title="About these documents" href="about.html" />
    <link rel="copyright" title="Copyright" href="copyright.html" />
    <link rel="top" title="Omics Pipe v1.0.16 Documentation" href="index.html" /> 
        <meta name="viewport" content="width=device-width, initial-scale=1">
  </head>
  <body>
    <div class="relbar-top">
        
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> &nbsp; &nbsp;</li>
    <li><a href="contents.html">Omics Pipe 1.0.16 Documentation</a> &raquo;</li>
 
      </ul>
    </div>
    </div>
  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="mirna-seq-count-based-modules">
<h1>miRNA-seq Count Based Modules<a class="headerlink" href="#mirna-seq-count-based-modules" title="Permalink to this headline">¶</a></h1>
<p>Modules available in the count-based miRNA-seq Pipeline.</p>
<div class="html-toggle section" id="cutadapt">
<h2>CutAdapt<a class="headerlink" href="#cutadapt" title="Permalink to this headline">¶</a></h2>
<span class="target" id="module-omics_pipe.modules.cutadapt_miRNA"></span><dl class="function">
<dt id="omics_pipe.modules.cutadapt_miRNA.cutadapt_miRNA">
<tt class="descclassname">omics_pipe.modules.cutadapt_miRNA.</tt><tt class="descname">cutadapt_miRNA</tt><big>(</big><em>sample</em>, <em>cutadapt_miRNA_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/cutadapt_miRNA.html#cutadapt_miRNA"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.cutadapt_miRNA.cutadapt_miRNA" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs Cutadapt to trim adapters from reads.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.fastq</dd>
<dt>output: </dt>
<dd>.fastq</dd>
<dt>citation: </dt>
<dd>Martin 2011. Cutadapt removes adapter sequences from high-throughput sequencing reads. EMBnet.journal 17: 10-12.</dd>
<dt>link: </dt>
<dd><a class="reference external" href="https://code.google.com/p/cutadapt/">https://code.google.com/p/cutadapt/</a></dd>
<dt>parameters from parameters file: </dt>
<dd>sample, RAW_DATA_DIR, ADAPTER, TRIMMED_DATA_PATH</dd>
</dl>
</dd></dl>

</div>
<div class="html-toggle section" id="fastq-length-filter">
<h2>Fastq Length Filter<a class="headerlink" href="#fastq-length-filter" title="Permalink to this headline">¶</a></h2>
<span class="target" id="module-omics_pipe.modules.fastq_length_filter_miRNA"></span><dl class="function">
<dt id="omics_pipe.modules.fastq_length_filter_miRNA.fastq_length_filter_miRNA">
<tt class="descclassname">omics_pipe.modules.fastq_length_filter_miRNA.</tt><tt class="descname">fastq_length_filter_miRNA</tt><big>(</big><em>sample</em>, <em>fastq_length_filter_miRNA_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/fastq_length_filter_miRNA.html#fastq_length_filter_miRNA"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.fastq_length_filter_miRNA.fastq_length_filter_miRNA" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs custom Python script to filter miRNA reads by length.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.fastq</dd>
<dt>output: </dt>
<dd>.fastq</dd>
<dt>parameters from parameter file: </dt>
<dd>TRIMMED_DATA_PATH:</dd>
</dl>
</dd></dl>

</div>
<div class="html-toggle section" id="module-omics_pipe.modules.fastqc">
<span id="fastqc"></span><h2>FASTQC<a class="headerlink" href="#module-omics_pipe.modules.fastqc" title="Permalink to this headline">¶</a></h2>
<dl class="function">
<dt id="omics_pipe.modules.fastqc.fastqc">
<tt class="descclassname">omics_pipe.modules.fastqc.</tt><tt class="descname">fastqc</tt><big>(</big><em>sample</em>, <em>fastqc_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/fastqc.html#fastqc"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.fastqc.fastqc" title="Permalink to this definition">¶</a></dt>
<dd><p>QC check of raw .fastq files using FASTQC.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.fastq file</dd>
<dt>output: </dt>
<dd>folder and zipped folder containing html, txt and image files</dd>
<dt>citation: </dt>
<dd>Babraham Bioinformatics</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://www.bioinformatics.babraham.ac.uk/projects/fastqc/">http://www.bioinformatics.babraham.ac.uk/projects/fastqc/</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">RAW_DATA_DIR:</p>
<p>QC_PATH:</p>
<p class="last">FASTQC_VERSION:</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="html-toggle section" id="star-aligner">
<h2>STAR Aligner<a class="headerlink" href="#star-aligner" title="Permalink to this headline">¶</a></h2>
<span class="target" id="module-omics_pipe.modules.star"></span><dl class="function">
<dt id="omics_pipe.modules.star.star">
<tt class="descclassname">omics_pipe.modules.star.</tt><tt class="descname">star</tt><big>(</big><em>sample</em>, <em>star_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/star.html#star"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.star.star" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs STAR to align .fastq files.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>.fastq file</dd>
<dt>output: </dt>
<dd>Aligned.out.bam</dd>
<dt>citation: </dt>
<dd><ol class="first last upperalpha simple">
<li>Dobin et al, Bioinformatics 2012; doi: 10.1093/bioinformatics/bts635 &#8220;STAR: ultrafast universal RNA-seq aligner&#8221;</li>
</ol>
</dd>
<dt>link: </dt>
<dd><a class="reference external" href="https://code.google.com/p/rna-star/">https://code.google.com/p/rna-star/</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">ENDS:</p>
<p>RAW_DATA_DIR:</p>
<p>STAR_INDEX:</p>
<p>STAR_OPTIONS:</p>
<p>STAR_RESULTS:</p>
<p>SAMTOOLS_VERSION:</p>
<p class="last">STAR_VERSION:</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="html-toggle section" id="htseq-count">
<h2>HTSEQ-count<a class="headerlink" href="#htseq-count" title="Permalink to this headline">¶</a></h2>
<span class="target" id="module-omics_pipe.modules.htseq"></span><dl class="function">
<dt id="omics_pipe.modules.htseq.htseq">
<tt class="descclassname">omics_pipe.modules.htseq.</tt><tt class="descname">htseq</tt><big>(</big><em>sample</em>, <em>htseq_flag</em><big>)</big><a class="reference internal" href="_modules/omics_pipe/modules/htseq.html#htseq"><span class="viewcode-link">[source]</span></a><a class="headerlink" href="#omics_pipe.modules.htseq.htseq" title="Permalink to this definition">¶</a></dt>
<dd><p>Runs htseq-count to get raw count data from alignments.</p>
<dl class="docutils">
<dt>input: </dt>
<dd>Aligned.out.sort.bam</dd>
<dt>output: </dt>
<dd>counts.txt</dd>
<dt>citation: </dt>
<dd>Simon Anders, EMBL</dd>
<dt>link: </dt>
<dd><a class="reference external" href="http://www-huber.embl.de/users/anders/HTSeq/doc/overview.html">http://www-huber.embl.de/users/anders/HTSeq/doc/overview.html</a></dd>
<dt>parameters from parameters file: </dt>
<dd><p class="first">STAR_RESULTS:</p>
<p>HTSEQ_OPTIONS:</p>
<p>REF_GENES:</p>
<p>HTSEQ_RESULTS:</p>
<p>TEMP_DIR:</p>
<p>SAMTOOLS_VERSION:</p>
<p class="last">BAM_FILE_NAME:</p>
</dd>
</dl>
</dd></dl>

</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar">
        <div class="sphinxsidebarwrapper">
        <p class="logo"><a href="contents.html" title="contents">
          <img class="logo" src="_static/logo.jpg" alt="Logo"/>
        </a></p><div class="sphinxlocaltoc">
    <h3><a href="contents.html">Page contents</a></h3>
    <ul>
<li><a class="reference internal" href="#">miRNA-seq Count Based Modules</a><ul>
<li><a class="reference internal" href="#cutadapt">CutAdapt</a></li>
<li><a class="reference internal" href="#fastq-length-filter">Fastq Length Filter</a></li>
<li><a class="reference internal" href="#module-omics_pipe.modules.fastqc">FASTQC</a></li>
<li><a class="reference internal" href="#star-aligner">STAR Aligner</a></li>
<li><a class="reference internal" href="#htseq-count">HTSEQ-count</a></li>
</ul>
</li>
</ul>

  </div>
  <h3>This Page</h3>
  <ul class="this-page-menu">
    <li><a href="_sources/miRNAseq_counts_modules.txt"
           rel="nofollow">Show Source</a></li>
  </ul>
<div id="searchbox" style="display: none">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" />
      <input type="submit" value="Go" />
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
    <p class="searchtip" style="font-size: 90%">
    Enter search terms or a module, class or function name.
    </p>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="relbar-bottom">
        
    <div class="related">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> &nbsp; &nbsp;</li>
    <li><a href="contents.html">Omics Pipe 1.0.16 Documentation</a> &raquo;</li>
 
      </ul>
    </div>
    </div>

    <div class="footer">
        &copy; <a href="copyright.html">Copyright</a> 2014, Kathleen Fisch, Ph.D..
      Last updated on Jul 09, 2014.
      Created using <a href="http://sphinx.pocoo.org/">Sphinx</a> 1.1.3.
    </div>
    <!-- cloud_sptheme 1.4 -->
  </body>
</html>