diff -r cdeaf44d10c9 django/http/__init__.py
--- a/django/http/__init__.py	Tue Apr 07 12:35:07 2009 -0400
+++ b/django/http/__init__.py	Tue Apr 07 13:18:33 2009 -0400
@@ -4,6 +4,8 @@
 from pprint import pformat
 from urllib import urlencode
 from urlparse import urljoin
+
+
 try:
     # The mod_python version is more efficient, so try importing it first.
     from mod_python.util import parse_qsl
@@ -21,8 +23,18 @@
 
 absolute_http_url_re = re.compile(r"^https?://", re.I)
 
-class Http404(Exception):
-    pass
+try:
+    # Try to import HTTPNotFound first
+    from webob.exc import HTTPNotFound
+except ImportError:
+    # We'll provide a fallback in case that webob isn't available
+    class Http404(Exception):
+        """default 404 error response"""
+        pass
+else:
+    class Http404(HTTPNotFound):
+        """WebOb-based 404 error response"""
+        pass
 
 class HttpRequest(object):
     """A basic HTTP request."""
