diff -r e08e6bbc8db0 django/forms/forms.py
--- a/django/forms/forms.py	Tue Jun 29 22:29:40 2010 -0400
+++ b/django/forms/forms.py	Wed Jun 30 09:51:34 2010 -0400
@@ -262,8 +262,8 @@
         self.cleaned_data = {}
         # If the form is permitted to be empty, and none of the form data has
         # changed from the initial data, short circuit any validation.
-        if self.empty_permitted and not self.has_changed():
-            return
+        #if self.empty_permitted and not self.has_changed():
+        #    return
         self._clean_fields()
         self._clean_form()
         self._post_clean()
