diff -r ab9695681e18 django/http/__init__.py
--- a/django/http/__init__.py	Wed Jun 30 09:51:34 2010 -0400
+++ b/django/http/__init__.py	Wed Jul 07 16:14:21 2010 -0400
@@ -343,6 +343,10 @@
             for key, value in self._headers.values()]) \
             + '\n\n' + self.content
 
+    def __len__(self):
+        """Return the length of the HTTP message."""
+        return int(len(self._container))
+
     def _convert_to_ascii(self, *values):
         """Converts all values to ascii strings."""
         for value in values:
