diff -r 7b0bf5422bd6 django/core/urlresolvers.py
--- a/django/core/urlresolvers.py	Wed Jul 07 16:14:21 2010 -0400
+++ b/django/core/urlresolvers.py	Wed Jul 07 16:15:39 2010 -0400
@@ -159,7 +159,7 @@
         lookups = MultiValueDict()
         namespaces = {}
         apps = {}
-        for pattern in reversed(self.url_patterns):
+        for pattern in reversed([item for item in iter(self.url_patterns)]):
             p_pattern = pattern.regex.pattern
             if p_pattern.startswith('^'):
                 p_pattern = p_pattern[1:]
