# HG changeset patch
# Parent ec50d5a91a788f26adbfb5c970d6788e4be24cda
diff --git a/django/core/urlresolvers.py b/django/core/urlresolvers.py
--- a/django/core/urlresolvers.py
+++ b/django/core/urlresolvers.py
@@ -288,14 +288,14 @@
         # lookup_view can be URL label, or dotted path, or callable, Any of
         # these can be passed in at the top, but callables are not friendly in
         # error messages.
-        m = getattr(lookup_view, '__module__', None)
-        n = getattr(lookup_view, '__name__', None)
-        if m is not None and n is not None:
-            lookup_view_s = "%s.%s" % (m, n)
-        else:
-            lookup_view_s = lookup_view
+        #m = getattr(lookup_view, '__module__', None)
+        #n = getattr(lookup_view, '__name__', None)
+        #if m is not None and n is not None:
+        #    lookup_view_s = "%s.%s" % (m, n)
+        #else:
+        #    lookup_view_s = lookup_view
         raise NoReverseMatch("Reverse for '%s' with arguments '%s' and keyword "
-                "arguments '%s' not found." % (lookup_view_s, args, kwargs))
+                "arguments '%s' not found." % (lookup_view, args, kwargs))
 
 def resolve(path, urlconf=None):
     if urlconf is None:
diff --git a/django/template/defaulttags.py b/django/template/defaulttags.py
--- a/django/template/defaulttags.py
+++ b/django/template/defaulttags.py
@@ -363,13 +363,13 @@
         # {% url ... as var %} construct in which cause return nothing.
         url = ''
         try:
-            url = reverse(self.view_name, args=args, kwargs=kwargs, current_app=context.current_app)
+            url = reverse(self.view_name, args=args, kwargs=kwargs)
         except NoReverseMatch, e:
             if settings.SETTINGS_MODULE:
                 project_name = settings.SETTINGS_MODULE.split('.')[0]
                 try:
                     url = reverse(project_name + '.' + self.view_name,
-                              args=args, kwargs=kwargs, current_app=context.current_app)
+                              args=args, kwargs=kwargs)
                 except NoReverseMatch:
                     if self.asvar is None:
                         # Re-raise the original exception, not the one with
