diff -r af6d3a70cc8c django/template/context.py
--- a/django/template/context.py	Sat Jan 30 04:51:30 2010 -0500
+++ b/django/template/context.py	Sat Jan 30 04:53:35 2010 -0500
@@ -12,7 +12,7 @@
     "pop() has been called more times than push()"
     pass
 
-class BaseContext(object):
+class BaseContext(dict):
     def __init__(self, dict_=None):
         dict_ = dict_ or {}
         self.dicts = [dict_]
